{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b5619054-d807-4a0d-bddb-084c218563e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import pickle\n",
    "import random\n",
    "import time\n",
    "import pandas as pd\n",
    "import webdataset as wds\n",
    "from tqdm.auto import tqdm\n",
    "import subprocess\n",
    "import csv \n",
    "from sklearn.model_selection import GridSearchCV, train_test_split\n",
    "from sklearn.preprocessing import LabelEncoder, label_binarize\n",
    "from sklearn.metrics import accuracy_score, roc_auc_score, recall_score\n",
    "from xgboost import XGBClassifier\n",
    "import xgboost as xgb\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "# from torchdata.datapipes.iter import FileLister, FileOpener\n",
    "# from functools import partial\n",
    "# import torch\n",
    "# from torch.utils.data import Dataset, DataLoader\n",
    "# import torch.nn as nn\n",
    "# import torch.nn.functional as F\n",
    "# import torch.optim as optim"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1d128f1-8b17-4724-b589-89e7f9a6db6e",
   "metadata": {},
   "source": [
    "# 1. Construct the multiomics dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc43b5cb-7047-457d-b270-8165686ba705",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 1(a). First, add the miRNA + gene expression dataset (primary tumor + normal samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bf91186-217b-4c8f-8811-8707da81ea1d",
   "metadata": {},
   "source": [
    "### Summary dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ef2a5f59-ce19-4d48-a245-e880be6e6b17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>gene_exp</th>\n",
       "      <th>snp</th>\n",
       "      <th>meth</th>\n",
       "      <th>n_omics_types</th>\n",
       "      <th>type</th>\n",
       "      <th>meth27k_dataset</th>\n",
       "      <th>meth450k_dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-D8-A146-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>df_BRCA_meth27</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-AQ-A0Y5-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_BRCA_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-C8-A1HI-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_BRCA_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A7-A0CD-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>df_BRCA_meth27</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-5L-AAT1-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_BRCA_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11886</th>\n",
       "      <td>TCGA-E2-A1IO-11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_BRCA_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11887</th>\n",
       "      <td>TCGA-BH-A0BF-11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_BRCA_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11888</th>\n",
       "      <td>TCGA-E2-A1IF-11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_BRCA_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11889</th>\n",
       "      <td>TCGA-BH-A1EY-11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_BRCA_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11890</th>\n",
       "      <td>TCGA-BH-A0DI-11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_BRCA_meth450</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11891 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Sample  miRNA  gene_exp  snp  meth  n_omics_types  \\\n",
       "0      TCGA-D8-A146-01      1         1    1     1              4   \n",
       "1      TCGA-AQ-A0Y5-01      1         1    1     1              4   \n",
       "2      TCGA-C8-A1HI-01      1         1    1     1              4   \n",
       "3      TCGA-A7-A0CD-01      1         1    1     1              4   \n",
       "4      TCGA-5L-AAT1-01      1         1    1     1              4   \n",
       "...                ...    ...       ...  ...   ...            ...   \n",
       "11886  TCGA-E2-A1IO-11      0         0    0     1              1   \n",
       "11887  TCGA-BH-A0BF-11      0         0    0     1              1   \n",
       "11888  TCGA-E2-A1IF-11      0         0    0     1              1   \n",
       "11889  TCGA-BH-A1EY-11      0         0    0     1              1   \n",
       "11890  TCGA-BH-A0DI-11      0         0    0     1              1   \n",
       "\n",
       "                      type meth27k_dataset meth450k_dataset  \n",
       "0            Primary Tumor  df_BRCA_meth27              NaN  \n",
       "1            Primary Tumor             NaN  df_BRCA_meth450  \n",
       "2            Primary Tumor             NaN  df_BRCA_meth450  \n",
       "3            Primary Tumor  df_BRCA_meth27              NaN  \n",
       "4            Primary Tumor             NaN  df_BRCA_meth450  \n",
       "...                    ...             ...              ...  \n",
       "11886  Solid Tissue Normal             NaN  df_BRCA_meth450  \n",
       "11887  Solid Tissue Normal             NaN  df_BRCA_meth450  \n",
       "11888  Solid Tissue Normal             NaN  df_BRCA_meth450  \n",
       "11889  Solid Tissue Normal             NaN  df_BRCA_meth450  \n",
       "11890  Solid Tissue Normal             NaN  df_BRCA_meth450  \n",
       "\n",
       "[11891 rows x 9 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"subset_primarytumorandnormal_with_omics_types.csv\", index_col = 0)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b0e7982c-358f-4d6c-9f5a-ba80ce3e9e0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>gene_exp</th>\n",
       "      <th>snp</th>\n",
       "      <th>meth</th>\n",
       "      <th>n_omics_types</th>\n",
       "      <th>type</th>\n",
       "      <th>meth27k_dataset</th>\n",
       "      <th>meth450k_dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-AQ-A0Y5-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_BRCA_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-C8-A1HI-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_BRCA_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-5L-AAT1-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_BRCA_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-BH-A0C0-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_BRCA_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-B6-A1KC-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_BRCA_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8462</th>\n",
       "      <td>TCGA-W5-AA31-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_CHOL_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8463</th>\n",
       "      <td>TCGA-W5-AA2U-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_CHOL_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8464</th>\n",
       "      <td>TCGA-ZU-A8S4-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_CHOL_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8465</th>\n",
       "      <td>TCGA-W5-AA2I-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_CHOL_meth450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8466</th>\n",
       "      <td>TCGA-W5-AA2H-01</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>df_CHOL_meth450</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8467 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Sample  miRNA  gene_exp  snp  meth  n_omics_types  \\\n",
       "0     TCGA-AQ-A0Y5-01      1         1    1     1              4   \n",
       "1     TCGA-C8-A1HI-01      1         1    1     1              4   \n",
       "2     TCGA-5L-AAT1-01      1         1    1     1              4   \n",
       "3     TCGA-BH-A0C0-01      1         1    1     1              4   \n",
       "4     TCGA-B6-A1KC-01      1         1    1     1              4   \n",
       "...               ...    ...       ...  ...   ...            ...   \n",
       "8462  TCGA-W5-AA31-01      1         1    1     1              4   \n",
       "8463  TCGA-W5-AA2U-01      1         1    1     1              4   \n",
       "8464  TCGA-ZU-A8S4-01      1         1    1     1              4   \n",
       "8465  TCGA-W5-AA2I-01      1         1    1     1              4   \n",
       "8466  TCGA-W5-AA2H-01      1         1    1     1              4   \n",
       "\n",
       "               type meth27k_dataset meth450k_dataset  \n",
       "0     Primary Tumor             NaN  df_BRCA_meth450  \n",
       "1     Primary Tumor             NaN  df_BRCA_meth450  \n",
       "2     Primary Tumor             NaN  df_BRCA_meth450  \n",
       "3     Primary Tumor             NaN  df_BRCA_meth450  \n",
       "4     Primary Tumor             NaN  df_BRCA_meth450  \n",
       "...             ...             ...              ...  \n",
       "8462  Primary Tumor             NaN  df_CHOL_meth450  \n",
       "8463  Primary Tumor             NaN  df_CHOL_meth450  \n",
       "8464  Primary Tumor             NaN  df_CHOL_meth450  \n",
       "8465  Primary Tumor             NaN  df_CHOL_meth450  \n",
       "8466  Primary Tumor             NaN  df_CHOL_meth450  \n",
       "\n",
       "[8467 rows x 9 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_subset = df[(df.miRNA == 1) & (df.gene_exp == 1) & (df.meth450k_dataset.notna())] # get samples with all 3 omics: gene expression + miRNA + methylation450k\n",
    "df_subset.reset_index(inplace=True, drop=True)\n",
    "df_subset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5eb19ecc-fc65-4eb1-b637-10ffe5e0c37f",
   "metadata": {},
   "source": [
    "### Gene expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bff7b7c4-fb08-43f6-a56d-479e8264f741",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5.780484016736349\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>ENSG00000000003.15</th>\n",
       "      <th>ENSG00000000005.6</th>\n",
       "      <th>ENSG00000000419.13</th>\n",
       "      <th>ENSG00000000457.14</th>\n",
       "      <th>ENSG00000000460.17</th>\n",
       "      <th>ENSG00000000938.13</th>\n",
       "      <th>ENSG00000000971.16</th>\n",
       "      <th>ENSG00000001036.14</th>\n",
       "      <th>ENSG00000001084.13</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000288665.1</th>\n",
       "      <th>ENSG00000288667.1</th>\n",
       "      <th>ENSG00000288669.1</th>\n",
       "      <th>ENSG00000288670.1</th>\n",
       "      <th>ENSG00000288671.1</th>\n",
       "      <th>ENSG00000288674.1</th>\n",
       "      <th>ENSG00000288675.1</th>\n",
       "      <th>type</th>\n",
       "      <th>stage</th>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-BRCA1</td>\n",
       "      <td>5.662037</td>\n",
       "      <td>3.376096</td>\n",
       "      <td>6.860140</td>\n",
       "      <td>4.400552</td>\n",
       "      <td>2.845169</td>\n",
       "      <td>3.579965</td>\n",
       "      <td>4.996181</td>\n",
       "      <td>5.614683</td>\n",
       "      <td>4.385652</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.533999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.058109</td>\n",
       "      <td>1.122739</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>TCGA-D8-A146-01A-31R-A115-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-BRCA2</td>\n",
       "      <td>3.703721</td>\n",
       "      <td>0.463099</td>\n",
       "      <td>7.086452</td>\n",
       "      <td>4.051007</td>\n",
       "      <td>2.426989</td>\n",
       "      <td>2.322361</td>\n",
       "      <td>4.943860</td>\n",
       "      <td>6.539078</td>\n",
       "      <td>4.046404</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.773448</td>\n",
       "      <td>0</td>\n",
       "      <td>0.054779</td>\n",
       "      <td>1.243182</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>TCGA-AQ-A0Y5-01A-11R-A14M-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-BRCA3</td>\n",
       "      <td>6.514515</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.805072</td>\n",
       "      <td>5.037264</td>\n",
       "      <td>4.043248</td>\n",
       "      <td>2.124130</td>\n",
       "      <td>2.993348</td>\n",
       "      <td>5.119912</td>\n",
       "      <td>3.409907</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.701710</td>\n",
       "      <td>0</td>\n",
       "      <td>0.093425</td>\n",
       "      <td>0.604261</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>TCGA-C8-A274-01A-11R-A16F-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-BRCA4</td>\n",
       "      <td>4.784917</td>\n",
       "      <td>2.328061</td>\n",
       "      <td>6.443071</td>\n",
       "      <td>4.374970</td>\n",
       "      <td>4.162790</td>\n",
       "      <td>3.122375</td>\n",
       "      <td>4.754674</td>\n",
       "      <td>5.608738</td>\n",
       "      <td>3.625691</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.966357</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.689075</td>\n",
       "      <td>0</td>\n",
       "      <td>0.059771</td>\n",
       "      <td>0.536351</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Stage IIA</td>\n",
       "      <td>TCGA-BH-A0BD-01A-11R-A034-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-BRCA5</td>\n",
       "      <td>4.251984</td>\n",
       "      <td>0.435415</td>\n",
       "      <td>6.178436</td>\n",
       "      <td>3.558464</td>\n",
       "      <td>2.589548</td>\n",
       "      <td>1.508581</td>\n",
       "      <td>3.722149</td>\n",
       "      <td>5.224044</td>\n",
       "      <td>3.242221</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.168221</td>\n",
       "      <td>0</td>\n",
       "      <td>0.023752</td>\n",
       "      <td>0.389126</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Stage IIB</td>\n",
       "      <td>TCGA-B6-A1KC-01B-11R-A157-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10814</th>\n",
       "      <td>TCGA-CHOL40</td>\n",
       "      <td>5.312015</td>\n",
       "      <td>0.054779</td>\n",
       "      <td>4.956693</td>\n",
       "      <td>1.847115</td>\n",
       "      <td>0.827900</td>\n",
       "      <td>2.303225</td>\n",
       "      <td>8.427157</td>\n",
       "      <td>4.806525</td>\n",
       "      <td>4.081399</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.263756</td>\n",
       "      <td>2.721504</td>\n",
       "      <td>0</td>\n",
       "      <td>0.008487</td>\n",
       "      <td>0.422233</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>TCGA-W5-AA30-11A-11R-A41I-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10815</th>\n",
       "      <td>TCGA-CHOL41</td>\n",
       "      <td>7.072792</td>\n",
       "      <td>0.070252</td>\n",
       "      <td>6.705809</td>\n",
       "      <td>2.356285</td>\n",
       "      <td>1.864176</td>\n",
       "      <td>2.305241</td>\n",
       "      <td>2.342270</td>\n",
       "      <td>6.360315</td>\n",
       "      <td>4.059762</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.773828</td>\n",
       "      <td>0</td>\n",
       "      <td>0.032806</td>\n",
       "      <td>1.390888</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>TCGA-W5-AA38-01A-11R-A41I-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10816</th>\n",
       "      <td>TCGA-CHOL42</td>\n",
       "      <td>2.132873</td>\n",
       "      <td>0.298893</td>\n",
       "      <td>6.025226</td>\n",
       "      <td>2.472904</td>\n",
       "      <td>1.310224</td>\n",
       "      <td>4.988394</td>\n",
       "      <td>3.466940</td>\n",
       "      <td>4.420954</td>\n",
       "      <td>3.570706</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1.080044</td>\n",
       "      <td>0.095182</td>\n",
       "      <td>2.120053</td>\n",
       "      <td>0</td>\n",
       "      <td>0.040261</td>\n",
       "      <td>1.388465</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>TCGA-W5-AA2H-01A-31R-A41I-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10817</th>\n",
       "      <td>TCGA-CHOL43</td>\n",
       "      <td>5.638529</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.587676</td>\n",
       "      <td>3.354960</td>\n",
       "      <td>2.290601</td>\n",
       "      <td>2.730901</td>\n",
       "      <td>6.821506</td>\n",
       "      <td>6.342717</td>\n",
       "      <td>3.480162</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.054779</td>\n",
       "      <td>4.356369</td>\n",
       "      <td>0</td>\n",
       "      <td>0.132643</td>\n",
       "      <td>1.277747</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>TCGA-W5-AA2O-01A-11R-A41I-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10818</th>\n",
       "      <td>TCGA-CHOL44</td>\n",
       "      <td>4.785148</td>\n",
       "      <td>0.090853</td>\n",
       "      <td>6.874128</td>\n",
       "      <td>4.296744</td>\n",
       "      <td>3.212849</td>\n",
       "      <td>2.201885</td>\n",
       "      <td>4.589218</td>\n",
       "      <td>6.126711</td>\n",
       "      <td>2.615722</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.829786</td>\n",
       "      <td>0</td>\n",
       "      <td>0.150040</td>\n",
       "      <td>1.498966</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Stage IVA</td>\n",
       "      <td>TCGA-W5-AA2W-01A-11R-A41I-07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10819 rows × 60664 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unnamed: 0  ENSG00000000003.15  ENSG00000000005.6  ENSG00000000419.13  \\\n",
       "0       TCGA-BRCA1            5.662037           3.376096            6.860140   \n",
       "1       TCGA-BRCA2            3.703721           0.463099            7.086452   \n",
       "2       TCGA-BRCA3            6.514515           0.000000            6.805072   \n",
       "3       TCGA-BRCA4            4.784917           2.328061            6.443071   \n",
       "4       TCGA-BRCA5            4.251984           0.435415            6.178436   \n",
       "...            ...                 ...                ...                 ...   \n",
       "10814  TCGA-CHOL40            5.312015           0.054779            4.956693   \n",
       "10815  TCGA-CHOL41            7.072792           0.070252            6.705809   \n",
       "10816  TCGA-CHOL42            2.132873           0.298893            6.025226   \n",
       "10817  TCGA-CHOL43            5.638529           0.000000            6.587676   \n",
       "10818  TCGA-CHOL44            4.785148           0.090853            6.874128   \n",
       "\n",
       "       ENSG00000000457.14  ENSG00000000460.17  ENSG00000000938.13  \\\n",
       "0                4.400552            2.845169            3.579965   \n",
       "1                4.051007            2.426989            2.322361   \n",
       "2                5.037264            4.043248            2.124130   \n",
       "3                4.374970            4.162790            3.122375   \n",
       "4                3.558464            2.589548            1.508581   \n",
       "...                   ...                 ...                 ...   \n",
       "10814            1.847115            0.827900            2.303225   \n",
       "10815            2.356285            1.864176            2.305241   \n",
       "10816            2.472904            1.310224            4.988394   \n",
       "10817            3.354960            2.290601            2.730901   \n",
       "10818            4.296744            3.212849            2.201885   \n",
       "\n",
       "       ENSG00000000971.16  ENSG00000001036.14  ENSG00000001084.13  ...  \\\n",
       "0                4.996181            5.614683            4.385652  ...   \n",
       "1                4.943860            6.539078            4.046404  ...   \n",
       "2                2.993348            5.119912            3.409907  ...   \n",
       "3                4.754674            5.608738            3.625691  ...   \n",
       "4                3.722149            5.224044            3.242221  ...   \n",
       "...                   ...                 ...                 ...  ...   \n",
       "10814            8.427157            4.806525            4.081399  ...   \n",
       "10815            2.342270            6.360315            4.059762  ...   \n",
       "10816            3.466940            4.420954            3.570706  ...   \n",
       "10817            6.821506            6.342717            3.480162  ...   \n",
       "10818            4.589218            6.126711            2.615722  ...   \n",
       "\n",
       "       ENSG00000288665.1  ENSG00000288667.1  ENSG00000288669.1  \\\n",
       "0                      0           0.000000           0.000000   \n",
       "1                      0           0.000000           0.000000   \n",
       "2                      0           0.000000           0.000000   \n",
       "3                      0           0.966357           0.000000   \n",
       "4                      0           0.000000           0.000000   \n",
       "...                  ...                ...                ...   \n",
       "10814                  0           0.000000           0.263756   \n",
       "10815                  0           0.000000           0.000000   \n",
       "10816                  0           1.080044           0.095182   \n",
       "10817                  0           0.000000           0.054779   \n",
       "10818                  0           0.000000           0.000000   \n",
       "\n",
       "       ENSG00000288670.1  ENSG00000288671.1  ENSG00000288674.1  \\\n",
       "0               3.533999                  0           0.058109   \n",
       "1               4.773448                  0           0.054779   \n",
       "2               4.701710                  0           0.093425   \n",
       "3               4.689075                  0           0.059771   \n",
       "4               5.168221                  0           0.023752   \n",
       "...                  ...                ...                ...   \n",
       "10814           2.721504                  0           0.008487   \n",
       "10815           3.773828                  0           0.032806   \n",
       "10816           2.120053                  0           0.040261   \n",
       "10817           4.356369                  0           0.132643   \n",
       "10818           5.829786                  0           0.150040   \n",
       "\n",
       "       ENSG00000288675.1                 type       stage  \\\n",
       "0               1.122739        Primary Tumor   Stage IIA   \n",
       "1               1.243182        Primary Tumor  Stage IIIA   \n",
       "2               0.604261        Primary Tumor   Stage IIA   \n",
       "3               0.536351        Primary Tumor   Stage IIA   \n",
       "4               0.389126        Primary Tumor   Stage IIB   \n",
       "...                  ...                  ...         ...   \n",
       "10814           0.422233  Solid Tissue Normal     Stage I   \n",
       "10815           1.390888        Primary Tumor     Stage I   \n",
       "10816           1.388465        Primary Tumor   Stage III   \n",
       "10817           1.277747        Primary Tumor     Stage I   \n",
       "10818           1.498966        Primary Tumor   Stage IVA   \n",
       "\n",
       "                            barcode  \n",
       "0      TCGA-D8-A146-01A-31R-A115-07  \n",
       "1      TCGA-AQ-A0Y5-01A-11R-A14M-07  \n",
       "2      TCGA-C8-A274-01A-11R-A16F-07  \n",
       "3      TCGA-BH-A0BD-01A-11R-A034-07  \n",
       "4      TCGA-B6-A1KC-01B-11R-A157-07  \n",
       "...                             ...  \n",
       "10814  TCGA-W5-AA30-11A-11R-A41I-07  \n",
       "10815  TCGA-W5-AA38-01A-11R-A41I-07  \n",
       "10816  TCGA-W5-AA2H-01A-31R-A41I-07  \n",
       "10817  TCGA-W5-AA2O-01A-11R-A41I-07  \n",
       "10818  TCGA-W5-AA2W-01A-11R-A41I-07  \n",
       "\n",
       "[10819 rows x 60664 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "df_gene_exp_full = pd.read_csv(\"for_yu_gene_expression.csv\")\n",
    "print((time.time() - start_time)/60)\n",
    "df_gene_exp_full"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c4515a7d-c9cd-4ef7-859b-c1013549348c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unnamed: 0\n",
      "type\n",
      "stage\n",
      "barcode\n"
     ]
    }
   ],
   "source": [
    "for col in df_gene_exp_full.columns:\n",
    "    if \"ENSG\" not in col:\n",
    "        print(col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "63c104e8-2ce2-413a-8279-313d4c938f47",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene_exp_full.drop(columns = ['Unnamed: 0', 'type', 'stage'], inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "73e4ddb2-4e25-469d-8a94-956113652376",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000003.15</th>\n",
       "      <th>ENSG00000000005.6</th>\n",
       "      <th>ENSG00000000419.13</th>\n",
       "      <th>ENSG00000000457.14</th>\n",
       "      <th>ENSG00000000460.17</th>\n",
       "      <th>ENSG00000000938.13</th>\n",
       "      <th>ENSG00000000971.16</th>\n",
       "      <th>ENSG00000001036.14</th>\n",
       "      <th>ENSG00000001084.13</th>\n",
       "      <th>ENSG00000001167.14</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000288662.1</th>\n",
       "      <th>ENSG00000288663.1</th>\n",
       "      <th>ENSG00000288665.1</th>\n",
       "      <th>ENSG00000288667.1</th>\n",
       "      <th>ENSG00000288669.1</th>\n",
       "      <th>ENSG00000288670.1</th>\n",
       "      <th>ENSG00000288671.1</th>\n",
       "      <th>ENSG00000288674.1</th>\n",
       "      <th>ENSG00000288675.1</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5.662037</td>\n",
       "      <td>3.376096</td>\n",
       "      <td>6.860140</td>\n",
       "      <td>4.400552</td>\n",
       "      <td>2.845169</td>\n",
       "      <td>3.579965</td>\n",
       "      <td>4.996181</td>\n",
       "      <td>5.614683</td>\n",
       "      <td>4.385652</td>\n",
       "      <td>5.843705</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.537346</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.533999</td>\n",
       "      <td>0</td>\n",
       "      <td>0.058109</td>\n",
       "      <td>1.122739</td>\n",
       "      <td>TCGA-D8-A146-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3.703721</td>\n",
       "      <td>0.463099</td>\n",
       "      <td>7.086452</td>\n",
       "      <td>4.051007</td>\n",
       "      <td>2.426989</td>\n",
       "      <td>2.322361</td>\n",
       "      <td>4.943860</td>\n",
       "      <td>6.539078</td>\n",
       "      <td>4.046404</td>\n",
       "      <td>5.711726</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.204767</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.773448</td>\n",
       "      <td>0</td>\n",
       "      <td>0.054779</td>\n",
       "      <td>1.243182</td>\n",
       "      <td>TCGA-AQ-A0Y5-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6.514515</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.805072</td>\n",
       "      <td>5.037264</td>\n",
       "      <td>4.043248</td>\n",
       "      <td>2.124130</td>\n",
       "      <td>2.993348</td>\n",
       "      <td>5.119912</td>\n",
       "      <td>3.409907</td>\n",
       "      <td>5.963742</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.412402</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.701710</td>\n",
       "      <td>0</td>\n",
       "      <td>0.093425</td>\n",
       "      <td>0.604261</td>\n",
       "      <td>TCGA-C8-A274-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.784917</td>\n",
       "      <td>2.328061</td>\n",
       "      <td>6.443071</td>\n",
       "      <td>4.374970</td>\n",
       "      <td>4.162790</td>\n",
       "      <td>3.122375</td>\n",
       "      <td>4.754674</td>\n",
       "      <td>5.608738</td>\n",
       "      <td>3.625691</td>\n",
       "      <td>5.678199</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.268075</td>\n",
       "      <td>0</td>\n",
       "      <td>0.966357</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.689075</td>\n",
       "      <td>0</td>\n",
       "      <td>0.059771</td>\n",
       "      <td>0.536351</td>\n",
       "      <td>TCGA-BH-A0BD-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4.251984</td>\n",
       "      <td>0.435415</td>\n",
       "      <td>6.178436</td>\n",
       "      <td>3.558464</td>\n",
       "      <td>2.589548</td>\n",
       "      <td>1.508581</td>\n",
       "      <td>3.722149</td>\n",
       "      <td>5.224044</td>\n",
       "      <td>3.242221</td>\n",
       "      <td>5.956145</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.434028</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.168221</td>\n",
       "      <td>0</td>\n",
       "      <td>0.023752</td>\n",
       "      <td>0.389126</td>\n",
       "      <td>TCGA-B6-A1KC-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10814</th>\n",
       "      <td>5.312015</td>\n",
       "      <td>0.054779</td>\n",
       "      <td>4.956693</td>\n",
       "      <td>1.847115</td>\n",
       "      <td>0.827900</td>\n",
       "      <td>2.303225</td>\n",
       "      <td>8.427157</td>\n",
       "      <td>4.806525</td>\n",
       "      <td>4.081399</td>\n",
       "      <td>2.794083</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.124593</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.263756</td>\n",
       "      <td>2.721504</td>\n",
       "      <td>0</td>\n",
       "      <td>0.008487</td>\n",
       "      <td>0.422233</td>\n",
       "      <td>TCGA-W5-AA30-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10815</th>\n",
       "      <td>7.072792</td>\n",
       "      <td>0.070252</td>\n",
       "      <td>6.705809</td>\n",
       "      <td>2.356285</td>\n",
       "      <td>1.864176</td>\n",
       "      <td>2.305241</td>\n",
       "      <td>2.342270</td>\n",
       "      <td>6.360315</td>\n",
       "      <td>4.059762</td>\n",
       "      <td>4.582773</td>\n",
       "      <td>...</td>\n",
       "      <td>1.15562</td>\n",
       "      <td>0.486148</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.773828</td>\n",
       "      <td>0</td>\n",
       "      <td>0.032806</td>\n",
       "      <td>1.390888</td>\n",
       "      <td>TCGA-W5-AA38-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10816</th>\n",
       "      <td>2.132873</td>\n",
       "      <td>0.298893</td>\n",
       "      <td>6.025226</td>\n",
       "      <td>2.472904</td>\n",
       "      <td>1.310224</td>\n",
       "      <td>4.988394</td>\n",
       "      <td>3.466940</td>\n",
       "      <td>4.420954</td>\n",
       "      <td>3.570706</td>\n",
       "      <td>4.070312</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.608525</td>\n",
       "      <td>0</td>\n",
       "      <td>1.080044</td>\n",
       "      <td>0.095182</td>\n",
       "      <td>2.120053</td>\n",
       "      <td>0</td>\n",
       "      <td>0.040261</td>\n",
       "      <td>1.388465</td>\n",
       "      <td>TCGA-W5-AA2H-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10817</th>\n",
       "      <td>5.638529</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.587676</td>\n",
       "      <td>3.354960</td>\n",
       "      <td>2.290601</td>\n",
       "      <td>2.730901</td>\n",
       "      <td>6.821506</td>\n",
       "      <td>6.342717</td>\n",
       "      <td>3.480162</td>\n",
       "      <td>3.630580</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.492007</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.054779</td>\n",
       "      <td>4.356369</td>\n",
       "      <td>0</td>\n",
       "      <td>0.132643</td>\n",
       "      <td>1.277747</td>\n",
       "      <td>TCGA-W5-AA2O-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10818</th>\n",
       "      <td>4.785148</td>\n",
       "      <td>0.090853</td>\n",
       "      <td>6.874128</td>\n",
       "      <td>4.296744</td>\n",
       "      <td>3.212849</td>\n",
       "      <td>2.201885</td>\n",
       "      <td>4.589218</td>\n",
       "      <td>6.126711</td>\n",
       "      <td>2.615722</td>\n",
       "      <td>5.301807</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.683202</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.829786</td>\n",
       "      <td>0</td>\n",
       "      <td>0.150040</td>\n",
       "      <td>1.498966</td>\n",
       "      <td>TCGA-W5-AA2W-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10819 rows × 60661 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       ENSG00000000003.15  ENSG00000000005.6  ENSG00000000419.13  \\\n",
       "0                5.662037           3.376096            6.860140   \n",
       "1                3.703721           0.463099            7.086452   \n",
       "2                6.514515           0.000000            6.805072   \n",
       "3                4.784917           2.328061            6.443071   \n",
       "4                4.251984           0.435415            6.178436   \n",
       "...                   ...                ...                 ...   \n",
       "10814            5.312015           0.054779            4.956693   \n",
       "10815            7.072792           0.070252            6.705809   \n",
       "10816            2.132873           0.298893            6.025226   \n",
       "10817            5.638529           0.000000            6.587676   \n",
       "10818            4.785148           0.090853            6.874128   \n",
       "\n",
       "       ENSG00000000457.14  ENSG00000000460.17  ENSG00000000938.13  \\\n",
       "0                4.400552            2.845169            3.579965   \n",
       "1                4.051007            2.426989            2.322361   \n",
       "2                5.037264            4.043248            2.124130   \n",
       "3                4.374970            4.162790            3.122375   \n",
       "4                3.558464            2.589548            1.508581   \n",
       "...                   ...                 ...                 ...   \n",
       "10814            1.847115            0.827900            2.303225   \n",
       "10815            2.356285            1.864176            2.305241   \n",
       "10816            2.472904            1.310224            4.988394   \n",
       "10817            3.354960            2.290601            2.730901   \n",
       "10818            4.296744            3.212849            2.201885   \n",
       "\n",
       "       ENSG00000000971.16  ENSG00000001036.14  ENSG00000001084.13  \\\n",
       "0                4.996181            5.614683            4.385652   \n",
       "1                4.943860            6.539078            4.046404   \n",
       "2                2.993348            5.119912            3.409907   \n",
       "3                4.754674            5.608738            3.625691   \n",
       "4                3.722149            5.224044            3.242221   \n",
       "...                   ...                 ...                 ...   \n",
       "10814            8.427157            4.806525            4.081399   \n",
       "10815            2.342270            6.360315            4.059762   \n",
       "10816            3.466940            4.420954            3.570706   \n",
       "10817            6.821506            6.342717            3.480162   \n",
       "10818            4.589218            6.126711            2.615722   \n",
       "\n",
       "       ENSG00000001167.14  ...  ENSG00000288662.1  ENSG00000288663.1  \\\n",
       "0                5.843705  ...            0.00000           0.537346   \n",
       "1                5.711726  ...            0.00000           0.204767   \n",
       "2                5.963742  ...            0.00000           0.412402   \n",
       "3                5.678199  ...            0.00000           0.268075   \n",
       "4                5.956145  ...            0.00000           0.434028   \n",
       "...                   ...  ...                ...                ...   \n",
       "10814            2.794083  ...            0.00000           0.124593   \n",
       "10815            4.582773  ...            1.15562           0.486148   \n",
       "10816            4.070312  ...            0.00000           0.608525   \n",
       "10817            3.630580  ...            0.00000           0.492007   \n",
       "10818            5.301807  ...            0.00000           0.683202   \n",
       "\n",
       "       ENSG00000288665.1  ENSG00000288667.1  ENSG00000288669.1  \\\n",
       "0                      0           0.000000           0.000000   \n",
       "1                      0           0.000000           0.000000   \n",
       "2                      0           0.000000           0.000000   \n",
       "3                      0           0.966357           0.000000   \n",
       "4                      0           0.000000           0.000000   \n",
       "...                  ...                ...                ...   \n",
       "10814                  0           0.000000           0.263756   \n",
       "10815                  0           0.000000           0.000000   \n",
       "10816                  0           1.080044           0.095182   \n",
       "10817                  0           0.000000           0.054779   \n",
       "10818                  0           0.000000           0.000000   \n",
       "\n",
       "       ENSG00000288670.1  ENSG00000288671.1  ENSG00000288674.1  \\\n",
       "0               3.533999                  0           0.058109   \n",
       "1               4.773448                  0           0.054779   \n",
       "2               4.701710                  0           0.093425   \n",
       "3               4.689075                  0           0.059771   \n",
       "4               5.168221                  0           0.023752   \n",
       "...                  ...                ...                ...   \n",
       "10814           2.721504                  0           0.008487   \n",
       "10815           3.773828                  0           0.032806   \n",
       "10816           2.120053                  0           0.040261   \n",
       "10817           4.356369                  0           0.132643   \n",
       "10818           5.829786                  0           0.150040   \n",
       "\n",
       "       ENSG00000288675.1        sample_id  \n",
       "0               1.122739  TCGA-D8-A146-01  \n",
       "1               1.243182  TCGA-AQ-A0Y5-01  \n",
       "2               0.604261  TCGA-C8-A274-01  \n",
       "3               0.536351  TCGA-BH-A0BD-01  \n",
       "4               0.389126  TCGA-B6-A1KC-01  \n",
       "...                  ...              ...  \n",
       "10814           0.422233  TCGA-W5-AA30-11  \n",
       "10815           1.390888  TCGA-W5-AA38-01  \n",
       "10816           1.388465  TCGA-W5-AA2H-01  \n",
       "10817           1.277747  TCGA-W5-AA2O-01  \n",
       "10818           1.498966  TCGA-W5-AA2W-01  \n",
       "\n",
       "[10819 rows x 60661 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract sample-level identifier (substring of the full barcode/sample ID)\n",
    "df_gene_exp_full['sample_id'] = df_gene_exp_full.barcode.str.split('-', n=4).str[:4].str.join('-').str[:-1]\n",
    "df_gene_exp_full = df_gene_exp_full.drop(columns = ['barcode'])\n",
    "df_gene_exp_full"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2e5f4d6a-f4dd-4c09-ae76-32ddcd162ffc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>type</th>\n",
       "      <th>ENSG00000000003.15</th>\n",
       "      <th>ENSG00000000005.6</th>\n",
       "      <th>ENSG00000000419.13</th>\n",
       "      <th>ENSG00000000457.14</th>\n",
       "      <th>ENSG00000000460.17</th>\n",
       "      <th>ENSG00000000938.13</th>\n",
       "      <th>ENSG00000000971.16</th>\n",
       "      <th>ENSG00000001036.14</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000288661.1</th>\n",
       "      <th>ENSG00000288662.1</th>\n",
       "      <th>ENSG00000288663.1</th>\n",
       "      <th>ENSG00000288665.1</th>\n",
       "      <th>ENSG00000288667.1</th>\n",
       "      <th>ENSG00000288669.1</th>\n",
       "      <th>ENSG00000288670.1</th>\n",
       "      <th>ENSG00000288671.1</th>\n",
       "      <th>ENSG00000288674.1</th>\n",
       "      <th>ENSG00000288675.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-AQ-A0Y5-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>3.703721</td>\n",
       "      <td>0.463099</td>\n",
       "      <td>7.086452</td>\n",
       "      <td>4.051007</td>\n",
       "      <td>2.426989</td>\n",
       "      <td>2.322361</td>\n",
       "      <td>4.943860</td>\n",
       "      <td>6.539078</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.204767</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.773448</td>\n",
       "      <td>0</td>\n",
       "      <td>0.054779</td>\n",
       "      <td>1.243182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-C8-A1HI-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>5.269714</td>\n",
       "      <td>0.289834</td>\n",
       "      <td>6.623920</td>\n",
       "      <td>4.429529</td>\n",
       "      <td>2.521980</td>\n",
       "      <td>2.750221</td>\n",
       "      <td>4.703101</td>\n",
       "      <td>6.267481</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.299245</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.119074</td>\n",
       "      <td>0</td>\n",
       "      <td>0.036468</td>\n",
       "      <td>0.709467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-5L-AAT1-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>4.899388</td>\n",
       "      <td>0.363227</td>\n",
       "      <td>6.779602</td>\n",
       "      <td>4.121861</td>\n",
       "      <td>2.128788</td>\n",
       "      <td>3.447725</td>\n",
       "      <td>4.697963</td>\n",
       "      <td>5.676846</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.579470</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.656840</td>\n",
       "      <td>0</td>\n",
       "      <td>0.020911</td>\n",
       "      <td>1.778671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-BH-A0C0-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>5.711255</td>\n",
       "      <td>0.268195</td>\n",
       "      <td>7.578444</td>\n",
       "      <td>3.913043</td>\n",
       "      <td>3.320687</td>\n",
       "      <td>4.947152</td>\n",
       "      <td>4.318447</td>\n",
       "      <td>6.316104</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.228973</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.725218</td>\n",
       "      <td>0</td>\n",
       "      <td>0.098285</td>\n",
       "      <td>0.344033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-B6-A1KC-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>4.251984</td>\n",
       "      <td>0.435415</td>\n",
       "      <td>6.178436</td>\n",
       "      <td>3.558464</td>\n",
       "      <td>2.589548</td>\n",
       "      <td>1.508581</td>\n",
       "      <td>3.722149</td>\n",
       "      <td>5.224044</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.434028</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.168221</td>\n",
       "      <td>0</td>\n",
       "      <td>0.023752</td>\n",
       "      <td>0.389126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8541</th>\n",
       "      <td>TCGA-W5-AA31-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>7.573606</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.811054</td>\n",
       "      <td>4.156073</td>\n",
       "      <td>2.804446</td>\n",
       "      <td>3.393155</td>\n",
       "      <td>7.001124</td>\n",
       "      <td>6.120097</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.713960</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.044884</td>\n",
       "      <td>3.955954</td>\n",
       "      <td>0</td>\n",
       "      <td>0.055473</td>\n",
       "      <td>0.806530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8542</th>\n",
       "      <td>TCGA-W5-AA2U-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>6.734234</td>\n",
       "      <td>0.061707</td>\n",
       "      <td>6.014679</td>\n",
       "      <td>3.355200</td>\n",
       "      <td>1.857424</td>\n",
       "      <td>2.760157</td>\n",
       "      <td>5.085195</td>\n",
       "      <td>5.275465</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.445621</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.939161</td>\n",
       "      <td>0</td>\n",
       "      <td>0.075327</td>\n",
       "      <td>0.468531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8543</th>\n",
       "      <td>TCGA-ZU-A8S4-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>5.233106</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.848891</td>\n",
       "      <td>2.474773</td>\n",
       "      <td>2.648627</td>\n",
       "      <td>4.104706</td>\n",
       "      <td>7.008132</td>\n",
       "      <td>7.073047</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.334225</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.441961</td>\n",
       "      <td>0</td>\n",
       "      <td>0.051442</td>\n",
       "      <td>1.265857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8544</th>\n",
       "      <td>TCGA-W5-AA2I-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>5.665532</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.152155</td>\n",
       "      <td>2.413811</td>\n",
       "      <td>1.658417</td>\n",
       "      <td>1.514198</td>\n",
       "      <td>3.354396</td>\n",
       "      <td>5.825127</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.457752</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.449258</td>\n",
       "      <td>0</td>\n",
       "      <td>0.039419</td>\n",
       "      <td>1.239887</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8545</th>\n",
       "      <td>TCGA-W5-AA2H-01</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>2.132873</td>\n",
       "      <td>0.298893</td>\n",
       "      <td>6.025226</td>\n",
       "      <td>2.472904</td>\n",
       "      <td>1.310224</td>\n",
       "      <td>4.988394</td>\n",
       "      <td>3.466940</td>\n",
       "      <td>4.420954</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.608525</td>\n",
       "      <td>0</td>\n",
       "      <td>1.080044</td>\n",
       "      <td>0.095182</td>\n",
       "      <td>2.120053</td>\n",
       "      <td>0</td>\n",
       "      <td>0.040261</td>\n",
       "      <td>1.388465</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8467 rows × 60662 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Sample           type  ENSG00000000003.15  ENSG00000000005.6  \\\n",
       "0     TCGA-AQ-A0Y5-01  Primary Tumor            3.703721           0.463099   \n",
       "1     TCGA-C8-A1HI-01  Primary Tumor            5.269714           0.289834   \n",
       "2     TCGA-5L-AAT1-01  Primary Tumor            4.899388           0.363227   \n",
       "3     TCGA-BH-A0C0-01  Primary Tumor            5.711255           0.268195   \n",
       "4     TCGA-B6-A1KC-01  Primary Tumor            4.251984           0.435415   \n",
       "...               ...            ...                 ...                ...   \n",
       "8541  TCGA-W5-AA31-01  Primary Tumor            7.573606           0.000000   \n",
       "8542  TCGA-W5-AA2U-01  Primary Tumor            6.734234           0.061707   \n",
       "8543  TCGA-ZU-A8S4-01  Primary Tumor            5.233106           0.000000   \n",
       "8544  TCGA-W5-AA2I-01  Primary Tumor            5.665532           0.000000   \n",
       "8545  TCGA-W5-AA2H-01  Primary Tumor            2.132873           0.298893   \n",
       "\n",
       "      ENSG00000000419.13  ENSG00000000457.14  ENSG00000000460.17  \\\n",
       "0               7.086452            4.051007            2.426989   \n",
       "1               6.623920            4.429529            2.521980   \n",
       "2               6.779602            4.121861            2.128788   \n",
       "3               7.578444            3.913043            3.320687   \n",
       "4               6.178436            3.558464            2.589548   \n",
       "...                  ...                 ...                 ...   \n",
       "8541            6.811054            4.156073            2.804446   \n",
       "8542            6.014679            3.355200            1.857424   \n",
       "8543            6.848891            2.474773            2.648627   \n",
       "8544            5.152155            2.413811            1.658417   \n",
       "8545            6.025226            2.472904            1.310224   \n",
       "\n",
       "      ENSG00000000938.13  ENSG00000000971.16  ENSG00000001036.14  ...  \\\n",
       "0               2.322361            4.943860            6.539078  ...   \n",
       "1               2.750221            4.703101            6.267481  ...   \n",
       "2               3.447725            4.697963            5.676846  ...   \n",
       "3               4.947152            4.318447            6.316104  ...   \n",
       "4               1.508581            3.722149            5.224044  ...   \n",
       "...                  ...                 ...                 ...  ...   \n",
       "8541            3.393155            7.001124            6.120097  ...   \n",
       "8542            2.760157            5.085195            5.275465  ...   \n",
       "8543            4.104706            7.008132            7.073047  ...   \n",
       "8544            1.514198            3.354396            5.825127  ...   \n",
       "8545            4.988394            3.466940            4.420954  ...   \n",
       "\n",
       "      ENSG00000288661.1  ENSG00000288662.1  ENSG00000288663.1  \\\n",
       "0                     0                0.0           0.204767   \n",
       "1                     0                0.0           0.299245   \n",
       "2                     0                0.0           0.579470   \n",
       "3                     0                0.0           0.228973   \n",
       "4                     0                0.0           0.434028   \n",
       "...                 ...                ...                ...   \n",
       "8541                  0                0.0           0.713960   \n",
       "8542                  0                0.0           0.445621   \n",
       "8543                  0                0.0           0.334225   \n",
       "8544                  0                0.0           0.457752   \n",
       "8545                  0                0.0           0.608525   \n",
       "\n",
       "      ENSG00000288665.1  ENSG00000288667.1  ENSG00000288669.1  \\\n",
       "0                     0           0.000000           0.000000   \n",
       "1                     0           0.000000           0.000000   \n",
       "2                     0           0.000000           0.000000   \n",
       "3                     0           0.000000           0.000000   \n",
       "4                     0           0.000000           0.000000   \n",
       "...                 ...                ...                ...   \n",
       "8541                  0           0.000000           0.044884   \n",
       "8542                  0           0.000000           0.000000   \n",
       "8543                  0           0.000000           0.000000   \n",
       "8544                  0           0.000000           0.000000   \n",
       "8545                  0           1.080044           0.095182   \n",
       "\n",
       "      ENSG00000288670.1  ENSG00000288671.1  ENSG00000288674.1  \\\n",
       "0              4.773448                  0           0.054779   \n",
       "1              4.119074                  0           0.036468   \n",
       "2              3.656840                  0           0.020911   \n",
       "3              4.725218                  0           0.098285   \n",
       "4              5.168221                  0           0.023752   \n",
       "...                 ...                ...                ...   \n",
       "8541           3.955954                  0           0.055473   \n",
       "8542           3.939161                  0           0.075327   \n",
       "8543           3.441961                  0           0.051442   \n",
       "8544           3.449258                  0           0.039419   \n",
       "8545           2.120053                  0           0.040261   \n",
       "\n",
       "      ENSG00000288675.1  \n",
       "0              1.243182  \n",
       "1              0.709467  \n",
       "2              1.778671  \n",
       "3              0.344033  \n",
       "4              0.389126  \n",
       "...                 ...  \n",
       "8541           0.806530  \n",
       "8542           0.468531  \n",
       "8543           1.265857  \n",
       "8544           1.239887  \n",
       "8545           1.388465  \n",
       "\n",
       "[8467 rows x 60662 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add gene exp data to dataframe\n",
    "df_multiomics = df_subset[['Sample', 'type']].merge(df_gene_exp_full, how = 'left', left_on = 'Sample', right_on = 'sample_id')\n",
    "df_multiomics = df_multiomics.drop(columns = ['sample_id'])\n",
    "df_multiomics.drop_duplicates(subset = ['Sample'], inplace= True)\n",
    "df_multiomics.shape # (8467, 60662)\n",
    "df_multiomics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc97bf23-1dc8-4165-a45b-c584687b3b7a",
   "metadata": {},
   "source": [
    "### miRNA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "59446fd5-556e-471d-a4bd-d799e37d6297",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>hsa-let-7a-1</th>\n",
       "      <th>hsa-let-7a-2</th>\n",
       "      <th>hsa-let-7a-3</th>\n",
       "      <th>hsa-let-7b</th>\n",
       "      <th>hsa-let-7c</th>\n",
       "      <th>hsa-let-7d</th>\n",
       "      <th>hsa-let-7e</th>\n",
       "      <th>hsa-let-7f-1</th>\n",
       "      <th>hsa-let-7f-2</th>\n",
       "      <th>...</th>\n",
       "      <th>hsa-mir-26b</th>\n",
       "      <th>hsa-mir-3116-1</th>\n",
       "      <th>hsa-mir-346</th>\n",
       "      <th>hsa-mir-4717</th>\n",
       "      <th>hsa-mir-548s</th>\n",
       "      <th>hsa-mir-5692c-1</th>\n",
       "      <th>hsa-mir-647</th>\n",
       "      <th>hsa-mir-6719</th>\n",
       "      <th>hsa-mir-6750</th>\n",
       "      <th>hsa-mir-6878</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-BRCA1</td>\n",
       "      <td>12.815167</td>\n",
       "      <td>12.816051</td>\n",
       "      <td>12.855016</td>\n",
       "      <td>14.730158</td>\n",
       "      <td>11.169628</td>\n",
       "      <td>8.859833</td>\n",
       "      <td>9.505782</td>\n",
       "      <td>10.680853</td>\n",
       "      <td>10.722648</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-BRCA2</td>\n",
       "      <td>12.968838</td>\n",
       "      <td>12.972606</td>\n",
       "      <td>12.972485</td>\n",
       "      <td>14.497080</td>\n",
       "      <td>11.869264</td>\n",
       "      <td>7.239808</td>\n",
       "      <td>9.593766</td>\n",
       "      <td>11.532029</td>\n",
       "      <td>11.479348</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-BRCA3</td>\n",
       "      <td>13.263951</td>\n",
       "      <td>13.247779</td>\n",
       "      <td>13.270085</td>\n",
       "      <td>13.844852</td>\n",
       "      <td>10.904361</td>\n",
       "      <td>9.067461</td>\n",
       "      <td>10.097007</td>\n",
       "      <td>12.528655</td>\n",
       "      <td>12.536764</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-BRCA4</td>\n",
       "      <td>12.871879</td>\n",
       "      <td>12.870939</td>\n",
       "      <td>12.896584</td>\n",
       "      <td>14.935364</td>\n",
       "      <td>9.644680</td>\n",
       "      <td>7.849328</td>\n",
       "      <td>8.665824</td>\n",
       "      <td>9.653458</td>\n",
       "      <td>9.730741</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-BRCA5</td>\n",
       "      <td>13.865815</td>\n",
       "      <td>13.861453</td>\n",
       "      <td>13.867042</td>\n",
       "      <td>14.606950</td>\n",
       "      <td>11.237808</td>\n",
       "      <td>8.316985</td>\n",
       "      <td>10.444484</td>\n",
       "      <td>12.448795</td>\n",
       "      <td>12.471484</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11085</th>\n",
       "      <td>TCGA-CHOL41</td>\n",
       "      <td>12.397355</td>\n",
       "      <td>12.410310</td>\n",
       "      <td>12.409395</td>\n",
       "      <td>12.764282</td>\n",
       "      <td>9.716770</td>\n",
       "      <td>8.727966</td>\n",
       "      <td>10.744548</td>\n",
       "      <td>11.404266</td>\n",
       "      <td>11.454670</td>\n",
       "      <td>...</td>\n",
       "      <td>8.120079</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.252878</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.252878</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11086</th>\n",
       "      <td>TCGA-CHOL42</td>\n",
       "      <td>12.271850</td>\n",
       "      <td>12.264286</td>\n",
       "      <td>12.282251</td>\n",
       "      <td>12.547766</td>\n",
       "      <td>9.808921</td>\n",
       "      <td>8.940528</td>\n",
       "      <td>10.275202</td>\n",
       "      <td>10.868141</td>\n",
       "      <td>10.869098</td>\n",
       "      <td>...</td>\n",
       "      <td>8.791045</td>\n",
       "      <td>0.235286</td>\n",
       "      <td>1.560000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.235286</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.235286</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11087</th>\n",
       "      <td>TCGA-CHOL43</td>\n",
       "      <td>13.402303</td>\n",
       "      <td>13.418145</td>\n",
       "      <td>13.407547</td>\n",
       "      <td>14.493636</td>\n",
       "      <td>10.488297</td>\n",
       "      <td>8.718217</td>\n",
       "      <td>9.983680</td>\n",
       "      <td>11.395059</td>\n",
       "      <td>11.392468</td>\n",
       "      <td>...</td>\n",
       "      <td>8.636674</td>\n",
       "      <td>0.326920</td>\n",
       "      <td>0.326920</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.818032</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11088</th>\n",
       "      <td>TCGA-CHOL44</td>\n",
       "      <td>12.914865</td>\n",
       "      <td>12.908972</td>\n",
       "      <td>12.913389</td>\n",
       "      <td>13.720097</td>\n",
       "      <td>9.832415</td>\n",
       "      <td>8.003953</td>\n",
       "      <td>10.085560</td>\n",
       "      <td>11.447787</td>\n",
       "      <td>11.492745</td>\n",
       "      <td>...</td>\n",
       "      <td>9.222545</td>\n",
       "      <td>0.142612</td>\n",
       "      <td>0.501421</td>\n",
       "      <td>0.142612</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.142612</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11089</th>\n",
       "      <td>TCGA-CHOL45</td>\n",
       "      <td>13.496989</td>\n",
       "      <td>13.506377</td>\n",
       "      <td>13.512535</td>\n",
       "      <td>14.241948</td>\n",
       "      <td>11.207995</td>\n",
       "      <td>9.435288</td>\n",
       "      <td>9.589841</td>\n",
       "      <td>12.162039</td>\n",
       "      <td>12.183739</td>\n",
       "      <td>...</td>\n",
       "      <td>10.325155</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.769251</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11090 rows × 1884 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unnamed: 0  hsa-let-7a-1  hsa-let-7a-2  hsa-let-7a-3  hsa-let-7b  \\\n",
       "0       TCGA-BRCA1     12.815167     12.816051     12.855016   14.730158   \n",
       "1       TCGA-BRCA2     12.968838     12.972606     12.972485   14.497080   \n",
       "2       TCGA-BRCA3     13.263951     13.247779     13.270085   13.844852   \n",
       "3       TCGA-BRCA4     12.871879     12.870939     12.896584   14.935364   \n",
       "4       TCGA-BRCA5     13.865815     13.861453     13.867042   14.606950   \n",
       "...            ...           ...           ...           ...         ...   \n",
       "11085  TCGA-CHOL41     12.397355     12.410310     12.409395   12.764282   \n",
       "11086  TCGA-CHOL42     12.271850     12.264286     12.282251   12.547766   \n",
       "11087  TCGA-CHOL43     13.402303     13.418145     13.407547   14.493636   \n",
       "11088  TCGA-CHOL44     12.914865     12.908972     12.913389   13.720097   \n",
       "11089  TCGA-CHOL45     13.496989     13.506377     13.512535   14.241948   \n",
       "\n",
       "       hsa-let-7c  hsa-let-7d  hsa-let-7e  hsa-let-7f-1  hsa-let-7f-2  ...  \\\n",
       "0       11.169628    8.859833    9.505782     10.680853     10.722648  ...   \n",
       "1       11.869264    7.239808    9.593766     11.532029     11.479348  ...   \n",
       "2       10.904361    9.067461   10.097007     12.528655     12.536764  ...   \n",
       "3        9.644680    7.849328    8.665824      9.653458      9.730741  ...   \n",
       "4       11.237808    8.316985   10.444484     12.448795     12.471484  ...   \n",
       "...           ...         ...         ...           ...           ...  ...   \n",
       "11085    9.716770    8.727966   10.744548     11.404266     11.454670  ...   \n",
       "11086    9.808921    8.940528   10.275202     10.868141     10.869098  ...   \n",
       "11087   10.488297    8.718217    9.983680     11.395059     11.392468  ...   \n",
       "11088    9.832415    8.003953   10.085560     11.447787     11.492745  ...   \n",
       "11089   11.207995    9.435288    9.589841     12.162039     12.183739  ...   \n",
       "\n",
       "       hsa-mir-26b  hsa-mir-3116-1  hsa-mir-346  hsa-mir-4717  hsa-mir-548s  \\\n",
       "0              NaN             NaN          NaN           NaN           NaN   \n",
       "1              NaN             NaN          NaN           NaN           NaN   \n",
       "2              NaN             NaN          NaN           NaN           NaN   \n",
       "3              NaN             NaN          NaN           NaN           NaN   \n",
       "4              NaN             NaN          NaN           NaN           NaN   \n",
       "...            ...             ...          ...           ...           ...   \n",
       "11085     8.120079        0.000000     0.252878      0.000000      0.000000   \n",
       "11086     8.791045        0.235286     1.560000      0.000000      0.235286   \n",
       "11087     8.636674        0.326920     0.326920      0.000000      0.818032   \n",
       "11088     9.222545        0.142612     0.501421      0.142612      0.000000   \n",
       "11089    10.325155        0.000000     0.000000      0.000000      1.769251   \n",
       "\n",
       "       hsa-mir-5692c-1  hsa-mir-647  hsa-mir-6719  hsa-mir-6750  hsa-mir-6878  \n",
       "0                  NaN          NaN           NaN           NaN           NaN  \n",
       "1                  NaN          NaN           NaN           NaN           NaN  \n",
       "2                  NaN          NaN           NaN           NaN           NaN  \n",
       "3                  NaN          NaN           NaN           NaN           NaN  \n",
       "4                  NaN          NaN           NaN           NaN           NaN  \n",
       "...                ...          ...           ...           ...           ...  \n",
       "11085         0.000000          0.0           0.0      0.252878           0.0  \n",
       "11086         0.000000          0.0           0.0      0.235286           0.0  \n",
       "11087         0.000000          0.0           0.0      0.000000           0.0  \n",
       "11088         0.142612          0.0           0.0      0.000000           0.0  \n",
       "11089         0.000000          0.0           0.0      0.000000           0.0  \n",
       "\n",
       "[11090 rows x 1884 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mirna = pd.read_csv('for_yu_miRNA.csv')\n",
    "df_mirna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "faf024b8-6787-4788-8187-37a40eff0ee3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unnamed: 0\n",
      "type\n",
      "Sample\n"
     ]
    }
   ],
   "source": [
    "for col in df_mirna.columns:\n",
    "    if \"hsa\" not in col:\n",
    "        print(col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cac3adee-1538-47dc-8351-69a409d65635",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hsa-let-7a-1</th>\n",
       "      <th>hsa-let-7a-2</th>\n",
       "      <th>hsa-let-7a-3</th>\n",
       "      <th>hsa-let-7b</th>\n",
       "      <th>hsa-let-7c</th>\n",
       "      <th>hsa-let-7d</th>\n",
       "      <th>hsa-let-7e</th>\n",
       "      <th>hsa-let-7f-1</th>\n",
       "      <th>hsa-let-7f-2</th>\n",
       "      <th>hsa-let-7g</th>\n",
       "      <th>...</th>\n",
       "      <th>hsa-mir-26b</th>\n",
       "      <th>hsa-mir-3116-1</th>\n",
       "      <th>hsa-mir-346</th>\n",
       "      <th>hsa-mir-4717</th>\n",
       "      <th>hsa-mir-548s</th>\n",
       "      <th>hsa-mir-5692c-1</th>\n",
       "      <th>hsa-mir-647</th>\n",
       "      <th>hsa-mir-6719</th>\n",
       "      <th>hsa-mir-6750</th>\n",
       "      <th>hsa-mir-6878</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12.815167</td>\n",
       "      <td>12.816051</td>\n",
       "      <td>12.855016</td>\n",
       "      <td>14.730158</td>\n",
       "      <td>11.169628</td>\n",
       "      <td>8.859833</td>\n",
       "      <td>9.505782</td>\n",
       "      <td>10.680853</td>\n",
       "      <td>10.722648</td>\n",
       "      <td>8.288114</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12.968838</td>\n",
       "      <td>12.972606</td>\n",
       "      <td>12.972485</td>\n",
       "      <td>14.497080</td>\n",
       "      <td>11.869264</td>\n",
       "      <td>7.239808</td>\n",
       "      <td>9.593766</td>\n",
       "      <td>11.532029</td>\n",
       "      <td>11.479348</td>\n",
       "      <td>8.007486</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13.263951</td>\n",
       "      <td>13.247779</td>\n",
       "      <td>13.270085</td>\n",
       "      <td>13.844852</td>\n",
       "      <td>10.904361</td>\n",
       "      <td>9.067461</td>\n",
       "      <td>10.097007</td>\n",
       "      <td>12.528655</td>\n",
       "      <td>12.536764</td>\n",
       "      <td>9.762910</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12.871879</td>\n",
       "      <td>12.870939</td>\n",
       "      <td>12.896584</td>\n",
       "      <td>14.935364</td>\n",
       "      <td>9.644680</td>\n",
       "      <td>7.849328</td>\n",
       "      <td>8.665824</td>\n",
       "      <td>9.653458</td>\n",
       "      <td>9.730741</td>\n",
       "      <td>8.959749</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13.865815</td>\n",
       "      <td>13.861453</td>\n",
       "      <td>13.867042</td>\n",
       "      <td>14.606950</td>\n",
       "      <td>11.237808</td>\n",
       "      <td>8.316985</td>\n",
       "      <td>10.444484</td>\n",
       "      <td>12.448795</td>\n",
       "      <td>12.471484</td>\n",
       "      <td>8.850126</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11085</th>\n",
       "      <td>12.397355</td>\n",
       "      <td>12.410310</td>\n",
       "      <td>12.409395</td>\n",
       "      <td>12.764282</td>\n",
       "      <td>9.716770</td>\n",
       "      <td>8.727966</td>\n",
       "      <td>10.744548</td>\n",
       "      <td>11.404266</td>\n",
       "      <td>11.454670</td>\n",
       "      <td>8.626717</td>\n",
       "      <td>...</td>\n",
       "      <td>8.120079</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.252878</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.252878</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11086</th>\n",
       "      <td>12.271850</td>\n",
       "      <td>12.264286</td>\n",
       "      <td>12.282251</td>\n",
       "      <td>12.547766</td>\n",
       "      <td>9.808921</td>\n",
       "      <td>8.940528</td>\n",
       "      <td>10.275202</td>\n",
       "      <td>10.868141</td>\n",
       "      <td>10.869098</td>\n",
       "      <td>8.997139</td>\n",
       "      <td>...</td>\n",
       "      <td>8.791045</td>\n",
       "      <td>0.235286</td>\n",
       "      <td>1.560000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.235286</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.235286</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11087</th>\n",
       "      <td>13.402303</td>\n",
       "      <td>13.418145</td>\n",
       "      <td>13.407547</td>\n",
       "      <td>14.493636</td>\n",
       "      <td>10.488297</td>\n",
       "      <td>8.718217</td>\n",
       "      <td>9.983680</td>\n",
       "      <td>11.395059</td>\n",
       "      <td>11.392468</td>\n",
       "      <td>9.127377</td>\n",
       "      <td>...</td>\n",
       "      <td>8.636674</td>\n",
       "      <td>0.326920</td>\n",
       "      <td>0.326920</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.818032</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11088</th>\n",
       "      <td>12.914865</td>\n",
       "      <td>12.908972</td>\n",
       "      <td>12.913389</td>\n",
       "      <td>13.720097</td>\n",
       "      <td>9.832415</td>\n",
       "      <td>8.003953</td>\n",
       "      <td>10.085560</td>\n",
       "      <td>11.447787</td>\n",
       "      <td>11.492745</td>\n",
       "      <td>8.792510</td>\n",
       "      <td>...</td>\n",
       "      <td>9.222545</td>\n",
       "      <td>0.142612</td>\n",
       "      <td>0.501421</td>\n",
       "      <td>0.142612</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.142612</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11089</th>\n",
       "      <td>13.496989</td>\n",
       "      <td>13.506377</td>\n",
       "      <td>13.512535</td>\n",
       "      <td>14.241948</td>\n",
       "      <td>11.207995</td>\n",
       "      <td>9.435288</td>\n",
       "      <td>9.589841</td>\n",
       "      <td>12.162039</td>\n",
       "      <td>12.183739</td>\n",
       "      <td>10.952019</td>\n",
       "      <td>...</td>\n",
       "      <td>10.325155</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.769251</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11090 rows × 1882 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       hsa-let-7a-1  hsa-let-7a-2  hsa-let-7a-3  hsa-let-7b  hsa-let-7c  \\\n",
       "0         12.815167     12.816051     12.855016   14.730158   11.169628   \n",
       "1         12.968838     12.972606     12.972485   14.497080   11.869264   \n",
       "2         13.263951     13.247779     13.270085   13.844852   10.904361   \n",
       "3         12.871879     12.870939     12.896584   14.935364    9.644680   \n",
       "4         13.865815     13.861453     13.867042   14.606950   11.237808   \n",
       "...             ...           ...           ...         ...         ...   \n",
       "11085     12.397355     12.410310     12.409395   12.764282    9.716770   \n",
       "11086     12.271850     12.264286     12.282251   12.547766    9.808921   \n",
       "11087     13.402303     13.418145     13.407547   14.493636   10.488297   \n",
       "11088     12.914865     12.908972     12.913389   13.720097    9.832415   \n",
       "11089     13.496989     13.506377     13.512535   14.241948   11.207995   \n",
       "\n",
       "       hsa-let-7d  hsa-let-7e  hsa-let-7f-1  hsa-let-7f-2  hsa-let-7g  ...  \\\n",
       "0        8.859833    9.505782     10.680853     10.722648    8.288114  ...   \n",
       "1        7.239808    9.593766     11.532029     11.479348    8.007486  ...   \n",
       "2        9.067461   10.097007     12.528655     12.536764    9.762910  ...   \n",
       "3        7.849328    8.665824      9.653458      9.730741    8.959749  ...   \n",
       "4        8.316985   10.444484     12.448795     12.471484    8.850126  ...   \n",
       "...           ...         ...           ...           ...         ...  ...   \n",
       "11085    8.727966   10.744548     11.404266     11.454670    8.626717  ...   \n",
       "11086    8.940528   10.275202     10.868141     10.869098    8.997139  ...   \n",
       "11087    8.718217    9.983680     11.395059     11.392468    9.127377  ...   \n",
       "11088    8.003953   10.085560     11.447787     11.492745    8.792510  ...   \n",
       "11089    9.435288    9.589841     12.162039     12.183739   10.952019  ...   \n",
       "\n",
       "       hsa-mir-26b  hsa-mir-3116-1  hsa-mir-346  hsa-mir-4717  hsa-mir-548s  \\\n",
       "0              NaN             NaN          NaN           NaN           NaN   \n",
       "1              NaN             NaN          NaN           NaN           NaN   \n",
       "2              NaN             NaN          NaN           NaN           NaN   \n",
       "3              NaN             NaN          NaN           NaN           NaN   \n",
       "4              NaN             NaN          NaN           NaN           NaN   \n",
       "...            ...             ...          ...           ...           ...   \n",
       "11085     8.120079        0.000000     0.252878      0.000000      0.000000   \n",
       "11086     8.791045        0.235286     1.560000      0.000000      0.235286   \n",
       "11087     8.636674        0.326920     0.326920      0.000000      0.818032   \n",
       "11088     9.222545        0.142612     0.501421      0.142612      0.000000   \n",
       "11089    10.325155        0.000000     0.000000      0.000000      1.769251   \n",
       "\n",
       "       hsa-mir-5692c-1  hsa-mir-647  hsa-mir-6719  hsa-mir-6750  hsa-mir-6878  \n",
       "0                  NaN          NaN           NaN           NaN           NaN  \n",
       "1                  NaN          NaN           NaN           NaN           NaN  \n",
       "2                  NaN          NaN           NaN           NaN           NaN  \n",
       "3                  NaN          NaN           NaN           NaN           NaN  \n",
       "4                  NaN          NaN           NaN           NaN           NaN  \n",
       "...                ...          ...           ...           ...           ...  \n",
       "11085         0.000000          0.0           0.0      0.252878           0.0  \n",
       "11086         0.000000          0.0           0.0      0.235286           0.0  \n",
       "11087         0.000000          0.0           0.0      0.000000           0.0  \n",
       "11088         0.142612          0.0           0.0      0.000000           0.0  \n",
       "11089         0.000000          0.0           0.0      0.000000           0.0  \n",
       "\n",
       "[11090 rows x 1882 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mirna.drop(columns = ['Unnamed: 0', 'type',], inplace = True)\n",
    "df_mirna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4a5aa75d-60e4-41a2-b389-7286b09e8dc1",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hsa-let-7a-1</th>\n",
       "      <th>hsa-let-7a-2</th>\n",
       "      <th>hsa-let-7a-3</th>\n",
       "      <th>hsa-let-7b</th>\n",
       "      <th>hsa-let-7c</th>\n",
       "      <th>hsa-let-7d</th>\n",
       "      <th>hsa-let-7e</th>\n",
       "      <th>hsa-let-7f-1</th>\n",
       "      <th>hsa-let-7f-2</th>\n",
       "      <th>hsa-let-7g</th>\n",
       "      <th>...</th>\n",
       "      <th>hsa-mir-3116-1</th>\n",
       "      <th>hsa-mir-346</th>\n",
       "      <th>hsa-mir-4717</th>\n",
       "      <th>hsa-mir-548s</th>\n",
       "      <th>hsa-mir-5692c-1</th>\n",
       "      <th>hsa-mir-647</th>\n",
       "      <th>hsa-mir-6719</th>\n",
       "      <th>hsa-mir-6750</th>\n",
       "      <th>hsa-mir-6878</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12.815167</td>\n",
       "      <td>12.816051</td>\n",
       "      <td>12.855016</td>\n",
       "      <td>14.730158</td>\n",
       "      <td>11.169628</td>\n",
       "      <td>8.859833</td>\n",
       "      <td>9.505782</td>\n",
       "      <td>10.680853</td>\n",
       "      <td>10.722648</td>\n",
       "      <td>8.288114</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-D8-A146-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12.968838</td>\n",
       "      <td>12.972606</td>\n",
       "      <td>12.972485</td>\n",
       "      <td>14.497080</td>\n",
       "      <td>11.869264</td>\n",
       "      <td>7.239808</td>\n",
       "      <td>9.593766</td>\n",
       "      <td>11.532029</td>\n",
       "      <td>11.479348</td>\n",
       "      <td>8.007486</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-AQ-A0Y5-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13.263951</td>\n",
       "      <td>13.247779</td>\n",
       "      <td>13.270085</td>\n",
       "      <td>13.844852</td>\n",
       "      <td>10.904361</td>\n",
       "      <td>9.067461</td>\n",
       "      <td>10.097007</td>\n",
       "      <td>12.528655</td>\n",
       "      <td>12.536764</td>\n",
       "      <td>9.762910</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-C8-A1HI-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12.871879</td>\n",
       "      <td>12.870939</td>\n",
       "      <td>12.896584</td>\n",
       "      <td>14.935364</td>\n",
       "      <td>9.644680</td>\n",
       "      <td>7.849328</td>\n",
       "      <td>8.665824</td>\n",
       "      <td>9.653458</td>\n",
       "      <td>9.730741</td>\n",
       "      <td>8.959749</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-A7-A0CD-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13.865815</td>\n",
       "      <td>13.861453</td>\n",
       "      <td>13.867042</td>\n",
       "      <td>14.606950</td>\n",
       "      <td>11.237808</td>\n",
       "      <td>8.316985</td>\n",
       "      <td>10.444484</td>\n",
       "      <td>12.448795</td>\n",
       "      <td>12.471484</td>\n",
       "      <td>8.850126</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TCGA-5L-AAT1-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11085</th>\n",
       "      <td>12.397355</td>\n",
       "      <td>12.410310</td>\n",
       "      <td>12.409395</td>\n",
       "      <td>12.764282</td>\n",
       "      <td>9.716770</td>\n",
       "      <td>8.727966</td>\n",
       "      <td>10.744548</td>\n",
       "      <td>11.404266</td>\n",
       "      <td>11.454670</td>\n",
       "      <td>8.626717</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.252878</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.252878</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-W5-AA31-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11086</th>\n",
       "      <td>12.271850</td>\n",
       "      <td>12.264286</td>\n",
       "      <td>12.282251</td>\n",
       "      <td>12.547766</td>\n",
       "      <td>9.808921</td>\n",
       "      <td>8.940528</td>\n",
       "      <td>10.275202</td>\n",
       "      <td>10.868141</td>\n",
       "      <td>10.869098</td>\n",
       "      <td>8.997139</td>\n",
       "      <td>...</td>\n",
       "      <td>0.235286</td>\n",
       "      <td>1.560000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.235286</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.235286</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-W5-AA2U-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11087</th>\n",
       "      <td>13.402303</td>\n",
       "      <td>13.418145</td>\n",
       "      <td>13.407547</td>\n",
       "      <td>14.493636</td>\n",
       "      <td>10.488297</td>\n",
       "      <td>8.718217</td>\n",
       "      <td>9.983680</td>\n",
       "      <td>11.395059</td>\n",
       "      <td>11.392468</td>\n",
       "      <td>9.127377</td>\n",
       "      <td>...</td>\n",
       "      <td>0.326920</td>\n",
       "      <td>0.326920</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.818032</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-ZU-A8S4-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11088</th>\n",
       "      <td>12.914865</td>\n",
       "      <td>12.908972</td>\n",
       "      <td>12.913389</td>\n",
       "      <td>13.720097</td>\n",
       "      <td>9.832415</td>\n",
       "      <td>8.003953</td>\n",
       "      <td>10.085560</td>\n",
       "      <td>11.447787</td>\n",
       "      <td>11.492745</td>\n",
       "      <td>8.792510</td>\n",
       "      <td>...</td>\n",
       "      <td>0.142612</td>\n",
       "      <td>0.501421</td>\n",
       "      <td>0.142612</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.142612</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-W5-AA2I-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11089</th>\n",
       "      <td>13.496989</td>\n",
       "      <td>13.506377</td>\n",
       "      <td>13.512535</td>\n",
       "      <td>14.241948</td>\n",
       "      <td>11.207995</td>\n",
       "      <td>9.435288</td>\n",
       "      <td>9.589841</td>\n",
       "      <td>12.162039</td>\n",
       "      <td>12.183739</td>\n",
       "      <td>10.952019</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.769251</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TCGA-W5-AA2H-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11090 rows × 1882 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       hsa-let-7a-1  hsa-let-7a-2  hsa-let-7a-3  hsa-let-7b  hsa-let-7c  \\\n",
       "0         12.815167     12.816051     12.855016   14.730158   11.169628   \n",
       "1         12.968838     12.972606     12.972485   14.497080   11.869264   \n",
       "2         13.263951     13.247779     13.270085   13.844852   10.904361   \n",
       "3         12.871879     12.870939     12.896584   14.935364    9.644680   \n",
       "4         13.865815     13.861453     13.867042   14.606950   11.237808   \n",
       "...             ...           ...           ...         ...         ...   \n",
       "11085     12.397355     12.410310     12.409395   12.764282    9.716770   \n",
       "11086     12.271850     12.264286     12.282251   12.547766    9.808921   \n",
       "11087     13.402303     13.418145     13.407547   14.493636   10.488297   \n",
       "11088     12.914865     12.908972     12.913389   13.720097    9.832415   \n",
       "11089     13.496989     13.506377     13.512535   14.241948   11.207995   \n",
       "\n",
       "       hsa-let-7d  hsa-let-7e  hsa-let-7f-1  hsa-let-7f-2  hsa-let-7g  ...  \\\n",
       "0        8.859833    9.505782     10.680853     10.722648    8.288114  ...   \n",
       "1        7.239808    9.593766     11.532029     11.479348    8.007486  ...   \n",
       "2        9.067461   10.097007     12.528655     12.536764    9.762910  ...   \n",
       "3        7.849328    8.665824      9.653458      9.730741    8.959749  ...   \n",
       "4        8.316985   10.444484     12.448795     12.471484    8.850126  ...   \n",
       "...           ...         ...           ...           ...         ...  ...   \n",
       "11085    8.727966   10.744548     11.404266     11.454670    8.626717  ...   \n",
       "11086    8.940528   10.275202     10.868141     10.869098    8.997139  ...   \n",
       "11087    8.718217    9.983680     11.395059     11.392468    9.127377  ...   \n",
       "11088    8.003953   10.085560     11.447787     11.492745    8.792510  ...   \n",
       "11089    9.435288    9.589841     12.162039     12.183739   10.952019  ...   \n",
       "\n",
       "       hsa-mir-3116-1  hsa-mir-346  hsa-mir-4717  hsa-mir-548s  \\\n",
       "0                 NaN          NaN           NaN           NaN   \n",
       "1                 NaN          NaN           NaN           NaN   \n",
       "2                 NaN          NaN           NaN           NaN   \n",
       "3                 NaN          NaN           NaN           NaN   \n",
       "4                 NaN          NaN           NaN           NaN   \n",
       "...               ...          ...           ...           ...   \n",
       "11085        0.000000     0.252878      0.000000      0.000000   \n",
       "11086        0.235286     1.560000      0.000000      0.235286   \n",
       "11087        0.326920     0.326920      0.000000      0.818032   \n",
       "11088        0.142612     0.501421      0.142612      0.000000   \n",
       "11089        0.000000     0.000000      0.000000      1.769251   \n",
       "\n",
       "       hsa-mir-5692c-1  hsa-mir-647  hsa-mir-6719  hsa-mir-6750  hsa-mir-6878  \\\n",
       "0                  NaN          NaN           NaN           NaN           NaN   \n",
       "1                  NaN          NaN           NaN           NaN           NaN   \n",
       "2                  NaN          NaN           NaN           NaN           NaN   \n",
       "3                  NaN          NaN           NaN           NaN           NaN   \n",
       "4                  NaN          NaN           NaN           NaN           NaN   \n",
       "...                ...          ...           ...           ...           ...   \n",
       "11085         0.000000          0.0           0.0      0.252878           0.0   \n",
       "11086         0.000000          0.0           0.0      0.235286           0.0   \n",
       "11087         0.000000          0.0           0.0      0.000000           0.0   \n",
       "11088         0.142612          0.0           0.0      0.000000           0.0   \n",
       "11089         0.000000          0.0           0.0      0.000000           0.0   \n",
       "\n",
       "             sample_id  \n",
       "0      TCGA-D8-A146-01  \n",
       "1      TCGA-AQ-A0Y5-01  \n",
       "2      TCGA-C8-A1HI-01  \n",
       "3      TCGA-A7-A0CD-01  \n",
       "4      TCGA-5L-AAT1-01  \n",
       "...                ...  \n",
       "11085  TCGA-W5-AA31-01  \n",
       "11086  TCGA-W5-AA2U-01  \n",
       "11087  TCGA-ZU-A8S4-01  \n",
       "11088  TCGA-W5-AA2I-01  \n",
       "11089  TCGA-W5-AA2H-01  \n",
       "\n",
       "[11090 rows x 1882 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract sample-level identifier (substring of the full barcode/sample ID)\n",
    "df_mirna['sample_id'] = df_mirna.Sample.str.split('-', n=4).str[:4].str.join('-').str[:-1]\n",
    "df_mirna = df_mirna.drop(columns = ['Sample'])\n",
    "df_mirna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1becbbb4-447f-4851-b309-4891dbfc6c18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8467, 62543)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add miRNA data to the multiomics dataframe\n",
    "df_multiomics = df_multiomics.merge(df_mirna, how = 'left', left_on = 'Sample', right_on = 'sample_id')\n",
    "df_multiomics = df_multiomics.drop(columns = ['sample_id'])\n",
    "df_multiomics.drop_duplicates(subset = ['Sample'], inplace= True)\n",
    "df_multiomics.shape # (8467, 62543)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0f318f57-b79f-4c74-8313-382008d2b70c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "type\n",
       "Primary Tumor          8099\n",
       "Solid Tissue Normal     368\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_multiomics.value_counts('type') # 8099 primary tumor, 368 solid tissue normal"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b892861d-dc88-4961-9783-2679c219dd00",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 1(b). Add methylomics to dataframe with miRNA and gene expression\n",
    "We will add a subset of columns from the methylomics data as we were unable to add all 450k methylation columns into our multiomics model.\n",
    "Later, when we are splitting the data to build and test our model, we will follow the same train/test/val split that we used to obtain a subset of methylation columns to prevent any information leakage."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4318a5ed-f433-40dd-aca7-8a0d9085de53",
   "metadata": {},
   "source": [
    "### Methylation data: distilled dataset with 593 features only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8963830c-1fb0-4843-8e79-ed2b79b136f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cg00657094</th>\n",
       "      <th>cg00889627</th>\n",
       "      <th>cg00026222</th>\n",
       "      <th>cg00028507</th>\n",
       "      <th>cg09509365</th>\n",
       "      <th>cg00000957</th>\n",
       "      <th>cg07319375</th>\n",
       "      <th>cg00547365</th>\n",
       "      <th>cg05814135</th>\n",
       "      <th>cg00101629</th>\n",
       "      <th>...</th>\n",
       "      <th>cg26914705</th>\n",
       "      <th>cg07679679</th>\n",
       "      <th>cg00058879</th>\n",
       "      <th>cg20957542</th>\n",
       "      <th>cg00921237</th>\n",
       "      <th>cg27166177</th>\n",
       "      <th>cg05729745</th>\n",
       "      <th>cg25081912</th>\n",
       "      <th>cg00062282</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.411621</td>\n",
       "      <td>0.621735</td>\n",
       "      <td>0.040575</td>\n",
       "      <td>0.161868</td>\n",
       "      <td>0.281257</td>\n",
       "      <td>0.777497</td>\n",
       "      <td>0.098279</td>\n",
       "      <td>0.617392</td>\n",
       "      <td>0.745925</td>\n",
       "      <td>0.325123</td>\n",
       "      <td>...</td>\n",
       "      <td>0.052327</td>\n",
       "      <td>0.380909</td>\n",
       "      <td>0.663547</td>\n",
       "      <td>0.747777</td>\n",
       "      <td>0.211242</td>\n",
       "      <td>0.114195</td>\n",
       "      <td>0.174401</td>\n",
       "      <td>0.816961</td>\n",
       "      <td>0.424337</td>\n",
       "      <td>TCGA-NF-A4X2-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.545119</td>\n",
       "      <td>0.340770</td>\n",
       "      <td>0.176231</td>\n",
       "      <td>0.331660</td>\n",
       "      <td>0.863144</td>\n",
       "      <td>0.844783</td>\n",
       "      <td>0.073168</td>\n",
       "      <td>0.545142</td>\n",
       "      <td>0.757941</td>\n",
       "      <td>0.861854</td>\n",
       "      <td>...</td>\n",
       "      <td>0.059471</td>\n",
       "      <td>0.335478</td>\n",
       "      <td>0.807092</td>\n",
       "      <td>0.636874</td>\n",
       "      <td>0.223941</td>\n",
       "      <td>0.125103</td>\n",
       "      <td>0.202899</td>\n",
       "      <td>0.848802</td>\n",
       "      <td>0.276662</td>\n",
       "      <td>TCGA-NG-A4VU-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.278215</td>\n",
       "      <td>0.400863</td>\n",
       "      <td>0.045977</td>\n",
       "      <td>0.267730</td>\n",
       "      <td>0.417079</td>\n",
       "      <td>0.688139</td>\n",
       "      <td>0.111705</td>\n",
       "      <td>0.649042</td>\n",
       "      <td>0.630707</td>\n",
       "      <td>0.856070</td>\n",
       "      <td>...</td>\n",
       "      <td>0.069232</td>\n",
       "      <td>0.557395</td>\n",
       "      <td>0.119348</td>\n",
       "      <td>0.588657</td>\n",
       "      <td>0.271522</td>\n",
       "      <td>0.152596</td>\n",
       "      <td>0.335041</td>\n",
       "      <td>0.846619</td>\n",
       "      <td>0.497601</td>\n",
       "      <td>TCGA-N8-A56S-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.292703</td>\n",
       "      <td>0.973566</td>\n",
       "      <td>0.056402</td>\n",
       "      <td>0.459386</td>\n",
       "      <td>0.781441</td>\n",
       "      <td>0.830256</td>\n",
       "      <td>0.080549</td>\n",
       "      <td>0.578502</td>\n",
       "      <td>0.838083</td>\n",
       "      <td>0.868675</td>\n",
       "      <td>...</td>\n",
       "      <td>0.046855</td>\n",
       "      <td>0.838569</td>\n",
       "      <td>0.835077</td>\n",
       "      <td>0.750678</td>\n",
       "      <td>0.320655</td>\n",
       "      <td>0.141215</td>\n",
       "      <td>0.150263</td>\n",
       "      <td>0.859772</td>\n",
       "      <td>0.483871</td>\n",
       "      <td>TCGA-N8-A4PL-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.583085</td>\n",
       "      <td>0.724319</td>\n",
       "      <td>0.024996</td>\n",
       "      <td>0.324168</td>\n",
       "      <td>0.720580</td>\n",
       "      <td>0.708335</td>\n",
       "      <td>0.099528</td>\n",
       "      <td>0.554873</td>\n",
       "      <td>0.686611</td>\n",
       "      <td>0.910438</td>\n",
       "      <td>...</td>\n",
       "      <td>0.044736</td>\n",
       "      <td>0.435550</td>\n",
       "      <td>0.112859</td>\n",
       "      <td>0.721100</td>\n",
       "      <td>0.684331</td>\n",
       "      <td>0.163656</td>\n",
       "      <td>0.317459</td>\n",
       "      <td>0.869584</td>\n",
       "      <td>0.282028</td>\n",
       "      <td>TCGA-N5-A4RO-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9781</th>\n",
       "      <td>0.469836</td>\n",
       "      <td>0.963279</td>\n",
       "      <td>0.033784</td>\n",
       "      <td>0.351272</td>\n",
       "      <td>0.880702</td>\n",
       "      <td>0.907370</td>\n",
       "      <td>0.143262</td>\n",
       "      <td>0.776855</td>\n",
       "      <td>0.946176</td>\n",
       "      <td>0.637099</td>\n",
       "      <td>...</td>\n",
       "      <td>0.067650</td>\n",
       "      <td>0.691867</td>\n",
       "      <td>0.096245</td>\n",
       "      <td>0.869565</td>\n",
       "      <td>0.296224</td>\n",
       "      <td>0.337493</td>\n",
       "      <td>0.308202</td>\n",
       "      <td>0.882717</td>\n",
       "      <td>0.614417</td>\n",
       "      <td>TCGA-KO-8403-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9782</th>\n",
       "      <td>0.568090</td>\n",
       "      <td>0.974412</td>\n",
       "      <td>0.035681</td>\n",
       "      <td>0.330773</td>\n",
       "      <td>0.881204</td>\n",
       "      <td>0.934874</td>\n",
       "      <td>0.136689</td>\n",
       "      <td>0.810328</td>\n",
       "      <td>0.928536</td>\n",
       "      <td>0.943398</td>\n",
       "      <td>...</td>\n",
       "      <td>0.041640</td>\n",
       "      <td>0.508179</td>\n",
       "      <td>0.188961</td>\n",
       "      <td>0.860149</td>\n",
       "      <td>0.602053</td>\n",
       "      <td>0.199547</td>\n",
       "      <td>0.277133</td>\n",
       "      <td>0.924013</td>\n",
       "      <td>0.644224</td>\n",
       "      <td>TCGA-KN-8434-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9783</th>\n",
       "      <td>0.885347</td>\n",
       "      <td>0.966679</td>\n",
       "      <td>0.034579</td>\n",
       "      <td>0.779984</td>\n",
       "      <td>0.849670</td>\n",
       "      <td>0.886904</td>\n",
       "      <td>0.329963</td>\n",
       "      <td>0.774311</td>\n",
       "      <td>0.740426</td>\n",
       "      <td>0.743007</td>\n",
       "      <td>...</td>\n",
       "      <td>0.069921</td>\n",
       "      <td>0.714920</td>\n",
       "      <td>0.113606</td>\n",
       "      <td>0.612634</td>\n",
       "      <td>0.877271</td>\n",
       "      <td>0.666925</td>\n",
       "      <td>0.646401</td>\n",
       "      <td>0.912015</td>\n",
       "      <td>0.487964</td>\n",
       "      <td>TCGA-KO-8405-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9784</th>\n",
       "      <td>0.962767</td>\n",
       "      <td>0.985107</td>\n",
       "      <td>0.058243</td>\n",
       "      <td>0.146255</td>\n",
       "      <td>0.904303</td>\n",
       "      <td>0.919828</td>\n",
       "      <td>0.114765</td>\n",
       "      <td>0.914389</td>\n",
       "      <td>0.959818</td>\n",
       "      <td>0.933331</td>\n",
       "      <td>...</td>\n",
       "      <td>0.060348</td>\n",
       "      <td>0.628638</td>\n",
       "      <td>0.059546</td>\n",
       "      <td>0.889165</td>\n",
       "      <td>0.196848</td>\n",
       "      <td>0.148803</td>\n",
       "      <td>0.364181</td>\n",
       "      <td>0.907754</td>\n",
       "      <td>0.773497</td>\n",
       "      <td>TCGA-KN-8436-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9785</th>\n",
       "      <td>0.884178</td>\n",
       "      <td>0.985759</td>\n",
       "      <td>0.056448</td>\n",
       "      <td>0.547191</td>\n",
       "      <td>0.897384</td>\n",
       "      <td>0.899272</td>\n",
       "      <td>0.162468</td>\n",
       "      <td>0.667443</td>\n",
       "      <td>0.920454</td>\n",
       "      <td>0.904984</td>\n",
       "      <td>...</td>\n",
       "      <td>0.056691</td>\n",
       "      <td>0.661029</td>\n",
       "      <td>0.088187</td>\n",
       "      <td>0.777222</td>\n",
       "      <td>0.595913</td>\n",
       "      <td>0.356155</td>\n",
       "      <td>0.416898</td>\n",
       "      <td>0.892235</td>\n",
       "      <td>0.584392</td>\n",
       "      <td>TCGA-KN-8437-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9786 rows × 594 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      cg00657094  cg00889627  cg00026222  cg00028507  cg09509365  cg00000957  \\\n",
       "0       0.411621    0.621735    0.040575    0.161868    0.281257    0.777497   \n",
       "1       0.545119    0.340770    0.176231    0.331660    0.863144    0.844783   \n",
       "2       0.278215    0.400863    0.045977    0.267730    0.417079    0.688139   \n",
       "3       0.292703    0.973566    0.056402    0.459386    0.781441    0.830256   \n",
       "4       0.583085    0.724319    0.024996    0.324168    0.720580    0.708335   \n",
       "...          ...         ...         ...         ...         ...         ...   \n",
       "9781    0.469836    0.963279    0.033784    0.351272    0.880702    0.907370   \n",
       "9782    0.568090    0.974412    0.035681    0.330773    0.881204    0.934874   \n",
       "9783    0.885347    0.966679    0.034579    0.779984    0.849670    0.886904   \n",
       "9784    0.962767    0.985107    0.058243    0.146255    0.904303    0.919828   \n",
       "9785    0.884178    0.985759    0.056448    0.547191    0.897384    0.899272   \n",
       "\n",
       "      cg07319375  cg00547365  cg05814135  cg00101629  ...  cg26914705  \\\n",
       "0       0.098279    0.617392    0.745925    0.325123  ...    0.052327   \n",
       "1       0.073168    0.545142    0.757941    0.861854  ...    0.059471   \n",
       "2       0.111705    0.649042    0.630707    0.856070  ...    0.069232   \n",
       "3       0.080549    0.578502    0.838083    0.868675  ...    0.046855   \n",
       "4       0.099528    0.554873    0.686611    0.910438  ...    0.044736   \n",
       "...          ...         ...         ...         ...  ...         ...   \n",
       "9781    0.143262    0.776855    0.946176    0.637099  ...    0.067650   \n",
       "9782    0.136689    0.810328    0.928536    0.943398  ...    0.041640   \n",
       "9783    0.329963    0.774311    0.740426    0.743007  ...    0.069921   \n",
       "9784    0.114765    0.914389    0.959818    0.933331  ...    0.060348   \n",
       "9785    0.162468    0.667443    0.920454    0.904984  ...    0.056691   \n",
       "\n",
       "      cg07679679  cg00058879  cg20957542  cg00921237  cg27166177  cg05729745  \\\n",
       "0       0.380909    0.663547    0.747777    0.211242    0.114195    0.174401   \n",
       "1       0.335478    0.807092    0.636874    0.223941    0.125103    0.202899   \n",
       "2       0.557395    0.119348    0.588657    0.271522    0.152596    0.335041   \n",
       "3       0.838569    0.835077    0.750678    0.320655    0.141215    0.150263   \n",
       "4       0.435550    0.112859    0.721100    0.684331    0.163656    0.317459   \n",
       "...          ...         ...         ...         ...         ...         ...   \n",
       "9781    0.691867    0.096245    0.869565    0.296224    0.337493    0.308202   \n",
       "9782    0.508179    0.188961    0.860149    0.602053    0.199547    0.277133   \n",
       "9783    0.714920    0.113606    0.612634    0.877271    0.666925    0.646401   \n",
       "9784    0.628638    0.059546    0.889165    0.196848    0.148803    0.364181   \n",
       "9785    0.661029    0.088187    0.777222    0.595913    0.356155    0.416898   \n",
       "\n",
       "      cg25081912  cg00062282        sample_id  \n",
       "0       0.816961    0.424337  TCGA-NF-A4X2-01  \n",
       "1       0.848802    0.276662  TCGA-NG-A4VU-01  \n",
       "2       0.846619    0.497601  TCGA-N8-A56S-01  \n",
       "3       0.859772    0.483871  TCGA-N8-A4PL-01  \n",
       "4       0.869584    0.282028  TCGA-N5-A4RO-01  \n",
       "...          ...         ...              ...  \n",
       "9781    0.882717    0.614417  TCGA-KO-8403-01  \n",
       "9782    0.924013    0.644224  TCGA-KN-8434-01  \n",
       "9783    0.912015    0.487964  TCGA-KO-8405-01  \n",
       "9784    0.907754    0.773497  TCGA-KN-8436-01  \n",
       "9785    0.892235    0.584392  TCGA-KN-8437-01  \n",
       "\n",
       "[9786 rows x 594 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meth = pd.read_csv(\"TCGA_meth_593cols.csv\", index_col=0)\n",
    "df_meth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "315a6cc9-c8ca-4f47-8b86-cb33c6301d47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Multiomics (miRNA + gene expression + methylation) dataframe shape: (8467, 63136)\n"
     ]
    }
   ],
   "source": [
    "##\n",
    "df_multiomics = df_multiomics.merge(df_meth, how = 'inner', left_on = 'Sample', right_on = 'sample_id')\n",
    "df_multiomics = df_multiomics.drop(columns = ['sample_id'])\n",
    "\n",
    "df_multiomics.drop_duplicates(subset = ['Sample'], inplace= True)\n",
    "\n",
    "print(f\"Multiomics (miRNA + gene expression + methylation) dataframe shape: {df_multiomics.shape}\") # (8467, 63136)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02e00d5e-5449-4f5a-9235-d8c2ee0b066e",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 2. Load External Validation file \n",
    "`df_external_validation`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3ec75af4-ffbb-43e2-a686-247c76d6c687",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.8570780158042908\n",
      "(1842, 63138)\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "\n",
    "## External validation files with 3 omics (includes only 593 methylation features as above)\n",
    "df_external_validation = pd.read_csv('external_validation/mirna_AND_gene_exp_AND_methylation_UPDATED_63138.csv.gz', \n",
    "                                     compression='gzip', sep=',', index_col=0)\n",
    "\n",
    "print((time.time() - start_time) / 60)\n",
    "print(df_external_validation.shape)\n",
    "# df_external_validation (1842, 63138)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbd78a66-bca4-4046-95d2-a17b1fc862b9",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 3. Preparing our data for modeling: Defining the Train/Val/Test sets\n",
    "We will use same train and hold-out sets as the ones we used to obtain the top methylation features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da8c9b2d-7c30-43e9-b0a1-c2dcb14e9f37",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### 3(a). Training set from TCGA (get the sample names from all datasets in `train_short` and `val_short` and merge)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b7042930-3a44-4a83-9d2d-dd35c2d53cfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['meth_450_BLCA_train.csv',\n",
       " 'meth_450_UCEC_train.csv',\n",
       " 'meth_450_LUSC_train.csv',\n",
       " 'meth_450_THCA_train.csv',\n",
       " 'meth_450_PAAD_train.csv',\n",
       " 'meth_450_CHOL_train.csv',\n",
       " 'meth_450_BRCA_train.csv',\n",
       " 'meth_450_READ_train.csv',\n",
       " 'meth_450_KIRP_train.csv',\n",
       " 'meth_450_ESCA_train.csv',\n",
       " 'meth_450_LUAD_train.csv',\n",
       " 'meth_450_COAD_train.csv',\n",
       " 'meth_450_HNSC_train.csv',\n",
       " 'meth_450_LIHC_train.csv',\n",
       " 'meth_450_KIRC_train.csv',\n",
       " 'meth_450_PRAD_train.csv',\n",
       " 'meth_450_UCEC_val.csv',\n",
       " 'meth_450_ESCA_val.csv',\n",
       " 'meth_450_HNSC_val.csv',\n",
       " 'meth_450_PRAD_val.csv',\n",
       " 'meth_450_THCA_val.csv',\n",
       " 'meth_450_KIRC_val.csv',\n",
       " 'meth_450_LIHC_val.csv',\n",
       " 'meth_450_BRCA_val.csv',\n",
       " 'meth_450_BLCA_val.csv',\n",
       " 'meth_450_READ_val.csv',\n",
       " 'meth_450_COAD_val.csv',\n",
       " 'meth_450_KIRP_val.csv',\n",
       " 'meth_450_CHOL_val.csv',\n",
       " 'meth_450_LUSC_val.csv',\n",
       " 'meth_450_LUAD_val.csv',\n",
       " 'meth_450_PAAD_val.csv']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_dirname = 'meth_imp_cols/450k_with_ids/train_short/'\n",
    "train_files = [f for f in os.listdir(train_dirname) if ('meth_450_' in f) and ('train.csv' in f)]\n",
    "\n",
    "val_dirname = 'meth_imp_cols/450k_with_ids/val_short/'\n",
    "val_files = [f for f in os.listdir(val_dirname) if ('meth_450_' in f) and ('val.csv' in f)]\n",
    "\n",
    "files = train_files + val_files\n",
    "print(len(files))\n",
    "files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "62b84336-fd2a-4fc2-84ba-cb0e888859d1",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 meth_450_BLCA_train.csv\n",
      "1 meth_450_UCEC_train.csv\n",
      "2 meth_450_LUSC_train.csv\n",
      "3 meth_450_THCA_train.csv\n",
      "4 meth_450_PAAD_train.csv\n",
      "5 meth_450_CHOL_train.csv\n",
      "6 meth_450_BRCA_train.csv\n",
      "7 meth_450_READ_train.csv\n",
      "8 meth_450_KIRP_train.csv\n",
      "9 meth_450_ESCA_train.csv\n",
      "10 meth_450_LUAD_train.csv\n",
      "11 meth_450_COAD_train.csv\n",
      "12 meth_450_HNSC_train.csv\n",
      "13 meth_450_LIHC_train.csv\n",
      "14 meth_450_KIRC_train.csv\n",
      "15 meth_450_PRAD_train.csv\n",
      "16 meth_450_UCEC_val.csv\n",
      "17 meth_450_ESCA_val.csv\n",
      "18 meth_450_HNSC_val.csv\n",
      "19 meth_450_PRAD_val.csv\n",
      "20 meth_450_THCA_val.csv\n",
      "21 meth_450_KIRC_val.csv\n",
      "22 meth_450_LIHC_val.csv\n",
      "23 meth_450_BRCA_val.csv\n",
      "24 meth_450_BLCA_val.csv\n",
      "25 meth_450_READ_val.csv\n",
      "26 meth_450_COAD_val.csv\n",
      "27 meth_450_KIRP_val.csv\n",
      "28 meth_450_CHOL_val.csv\n",
      "29 meth_450_LUSC_val.csv\n",
      "30 meth_450_LUAD_val.csv\n",
      "31 meth_450_PAAD_val.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-CU-A3KJ-01A-11D-A21B-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-C4-A0F7-01A-11D-A10W-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-K4-A3WU-01B-11D-A23O-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-DK-A3IU-01A-11D-A211-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-XF-A9SU-01A-31D-A392-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5210</th>\n",
       "      <td>TCGA-FZ-5924-11A-01D-1744-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5211</th>\n",
       "      <td>TCGA-XD-AAUL-01A-21D-A398-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5212</th>\n",
       "      <td>TCGA-IB-7887-01A-11D-2157-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5213</th>\n",
       "      <td>TCGA-HZ-A77O-01A-11D-A33U-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5214</th>\n",
       "      <td>TCGA-LB-A8F3-01A-11D-A368-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5215 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            sample\n",
       "0     TCGA-CU-A3KJ-01A-11D-A21B-05\n",
       "1     TCGA-C4-A0F7-01A-11D-A10W-05\n",
       "2     TCGA-K4-A3WU-01B-11D-A23O-05\n",
       "3     TCGA-DK-A3IU-01A-11D-A211-05\n",
       "4     TCGA-XF-A9SU-01A-31D-A392-05\n",
       "...                            ...\n",
       "5210  TCGA-FZ-5924-11A-01D-1744-05\n",
       "5211  TCGA-XD-AAUL-01A-21D-A398-05\n",
       "5212  TCGA-IB-7887-01A-11D-2157-05\n",
       "5213  TCGA-HZ-A77O-01A-11D-A33U-05\n",
       "5214  TCGA-LB-A8F3-01A-11D-A368-05\n",
       "\n",
       "[5215 rows x 1 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = pd.DataFrame(columns = ['sample'])\n",
    "for f_index, f in enumerate(files):\n",
    "    print(f_index, f)\n",
    "    \n",
    "    if '_val.csv' in f:\n",
    "        dirname = val_dirname\n",
    "    elif '_train.csv' in f:\n",
    "        dirname = train_dirname\n",
    "    temp = pd.read_csv(dirname + f, usecols = ['sample'])\n",
    "    samples = pd.concat([samples, temp])\n",
    "    samples.reset_index(drop=True, inplace=True)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7b026142-ba3f-4cdf-bb18-1e0e107bc043",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-CU-A3KJ-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-C4-A0F7-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-K4-A3WU-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-DK-A3IU-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-XF-A9SU-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5210</th>\n",
       "      <td>TCGA-FZ-5924-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5211</th>\n",
       "      <td>TCGA-XD-AAUL-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5212</th>\n",
       "      <td>TCGA-IB-7887-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5213</th>\n",
       "      <td>TCGA-HZ-A77O-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5214</th>\n",
       "      <td>TCGA-LB-A8F3-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5215 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample_id\n",
       "0     TCGA-CU-A3KJ-01\n",
       "1     TCGA-C4-A0F7-01\n",
       "2     TCGA-K4-A3WU-01\n",
       "3     TCGA-DK-A3IU-01\n",
       "4     TCGA-XF-A9SU-01\n",
       "...               ...\n",
       "5210  TCGA-FZ-5924-11\n",
       "5211  TCGA-XD-AAUL-01\n",
       "5212  TCGA-IB-7887-01\n",
       "5213  TCGA-HZ-A77O-01\n",
       "5214  TCGA-LB-A8F3-01\n",
       "\n",
       "[5215 rows x 1 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# extract sample-level identifiers (substring of full sample IDs)\n",
    "samples['sample_id'] = samples['sample'].str.split('-', n=4).str[:4].str.join('-').str[:-1]\n",
    "samples = samples.drop(columns = ['sample'])\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "923f6994-7c68-4040-b735-d924ee528c12",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4863, 63135)\n"
     ]
    }
   ],
   "source": [
    "df = df_multiomics.copy()\n",
    "df['target'] =  (df['type'] == 'Primary Tumor').astype(int)\n",
    "df = df.drop(columns=['type'])\n",
    "\n",
    "train_df = samples.merge(df, how = 'inner', left_on='sample_id', right_on='Sample')\n",
    "\n",
    "# separate features and target\n",
    "X_train_all = train_df.drop(columns=['Sample', 'target'])\n",
    "y_train = train_df['target']\n",
    "\n",
    "print(X_train_all.shape) # 63135 (= 63134 features columns + 1 sample id column)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58c62007-dd5a-4c2a-90dd-fc9f783693e5",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### 3(b). Hold-out test set from TCGA (get the sample names from all datasets in `test_short` and merge)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8620a87f-ce06-41af-b726-070d5abe71e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['meth_450_PRAD_test.csv',\n",
       " 'meth_450_LUAD_test.csv',\n",
       " 'meth_450_LUSC_test.csv',\n",
       " 'meth_450_KIRC_test.csv',\n",
       " 'meth_450_KIRP_test.csv',\n",
       " 'meth_450_LIHC_test.csv',\n",
       " 'meth_450_HNSC_test.csv',\n",
       " 'meth_450_READ_test.csv',\n",
       " 'meth_450_CHOL_test.csv',\n",
       " 'meth_450_BLCA_test.csv',\n",
       " 'meth_450_PAAD_test.csv',\n",
       " 'meth_450_UCEC_test.csv',\n",
       " 'meth_450_BRCA_test.csv',\n",
       " 'meth_450_ESCA_test.csv',\n",
       " 'meth_450_THCA_test.csv',\n",
       " 'meth_450_COAD_test.csv']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_dirname = 'meth_imp_cols/450k_with_ids/test_short/'\n",
    "files = [f for f in os.listdir(test_dirname) if ('meth_450_' in f) and ('test.csv' in f)]\n",
    "\n",
    "print(len(files))\n",
    "files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "21ec064c-f699-434b-8098-5d98c4e56a7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 meth_450_PRAD_test.csv\n",
      "1 meth_450_LUAD_test.csv\n",
      "2 meth_450_LUSC_test.csv\n",
      "3 meth_450_KIRC_test.csv\n",
      "4 meth_450_KIRP_test.csv\n",
      "5 meth_450_LIHC_test.csv\n",
      "6 meth_450_HNSC_test.csv\n",
      "7 meth_450_READ_test.csv\n",
      "8 meth_450_CHOL_test.csv\n",
      "9 meth_450_BLCA_test.csv\n",
      "10 meth_450_PAAD_test.csv\n",
      "11 meth_450_UCEC_test.csv\n",
      "12 meth_450_BRCA_test.csv\n",
      "13 meth_450_ESCA_test.csv\n",
      "14 meth_450_THCA_test.csv\n",
      "15 meth_450_COAD_test.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-G9-6338-01A-12D-1963-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-EJ-7781-11A-01D-2116-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-HC-A8D1-01A-11D-A365-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-EJ-7315-01A-31D-2116-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-VP-A878-01A-31D-A34V-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1329</th>\n",
       "      <td>TCGA-AY-A71X-01A-12D-A36Y-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1330</th>\n",
       "      <td>TCGA-CM-5860-01A-01D-1651-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1331</th>\n",
       "      <td>TCGA-CM-6674-01A-11D-1837-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1332</th>\n",
       "      <td>TCGA-G4-6294-01A-11D-1772-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1333</th>\n",
       "      <td>TCGA-G4-6310-01A-11D-1721-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1334 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            sample\n",
       "0     TCGA-G9-6338-01A-12D-1963-05\n",
       "1     TCGA-EJ-7781-11A-01D-2116-05\n",
       "2     TCGA-HC-A8D1-01A-11D-A365-05\n",
       "3     TCGA-EJ-7315-01A-31D-2116-05\n",
       "4     TCGA-VP-A878-01A-31D-A34V-05\n",
       "...                            ...\n",
       "1329  TCGA-AY-A71X-01A-12D-A36Y-05\n",
       "1330  TCGA-CM-5860-01A-01D-1651-05\n",
       "1331  TCGA-CM-6674-01A-11D-1837-05\n",
       "1332  TCGA-G4-6294-01A-11D-1772-05\n",
       "1333  TCGA-G4-6310-01A-11D-1721-05\n",
       "\n",
       "[1334 rows x 1 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = pd.DataFrame(columns = ['sample'])\n",
    "for f_index, f in enumerate(files):\n",
    "    print(f_index, f)\n",
    "    if '_test.csv' in f:\n",
    "        dirname = test_dirname\n",
    "    temp = pd.read_csv(dirname + f, usecols = ['sample'])\n",
    "    samples = pd.concat([samples, temp])\n",
    "    samples.reset_index(drop=True, inplace=True)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e5e02ac7-065d-4d86-a027-dbafc7f3187b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-G9-6338-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-EJ-7781-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-HC-A8D1-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-EJ-7315-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-VP-A878-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1329</th>\n",
       "      <td>TCGA-AY-A71X-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1330</th>\n",
       "      <td>TCGA-CM-5860-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1331</th>\n",
       "      <td>TCGA-CM-6674-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1332</th>\n",
       "      <td>TCGA-G4-6294-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1333</th>\n",
       "      <td>TCGA-G4-6310-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1334 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sample_id\n",
       "0     TCGA-G9-6338-01\n",
       "1     TCGA-EJ-7781-11\n",
       "2     TCGA-HC-A8D1-01\n",
       "3     TCGA-EJ-7315-01\n",
       "4     TCGA-VP-A878-01\n",
       "...               ...\n",
       "1329  TCGA-AY-A71X-01\n",
       "1330  TCGA-CM-5860-01\n",
       "1331  TCGA-CM-6674-01\n",
       "1332  TCGA-G4-6294-01\n",
       "1333  TCGA-G4-6310-01\n",
       "\n",
       "[1334 rows x 1 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples['sample_id'] = samples['sample'].str.split('-', n=4).str[:4].str.join('-').str[:-1]\n",
    "samples = samples.drop(columns = ['sample'])\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5f5c088a-50fb-4da3-ad3a-6c989b613b31",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1239, 63135)\n"
     ]
    }
   ],
   "source": [
    "df = df_multiomics.copy()\n",
    "df['target'] =  (df['type'] == 'Primary Tumor').astype(int)\n",
    "df = df.drop(columns=['type'])\n",
    "\n",
    "test_df = samples.merge(df, how = 'inner', left_on='sample_id', right_on='Sample')\n",
    "\n",
    "# separate features and target\n",
    "X_test_all = test_df.drop(columns=['Sample', 'target'])\n",
    "y_test = test_df['target']\n",
    "\n",
    "print(X_test_all.shape) # 63135 (= 63134 features columns + 1 sample id column)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8f5b982-059e-4f62-bcee-0862fe250238",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 3(c). External test set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1e302856-e112-4a5e-8fb8-a263e03883d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1842, 63137)\n"
     ]
    }
   ],
   "source": [
    "# external test set\n",
    "df_ext_test = df_external_validation.copy()\n",
    "df_ext_test['target'] = (df_ext_test['Type'] == 'Primary Tumor').astype(int)\n",
    "df_ext_test = df_ext_test.drop(columns=['Type'])\n",
    "\n",
    "# separate features and target\n",
    "X_ext_test_all = df_ext_test.drop(columns=['target'])\n",
    "y_ext_test = df_ext_test['target']\n",
    "print(X_ext_test_all.shape) # 63137 (= 63134 features + 1 sample id + 1 site + 1 stage)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39ed498b-73cb-45ff-9d91-f83b605e743c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Summary of part 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b2e8a414-4830-4708-8684-8f374d603ac8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4863, 63135) (4863,)\n",
      "(1239, 63135) (1239,)\n",
      "(1842, 63137) (1842,)\n"
     ]
    }
   ],
   "source": [
    "print(X_train_all.shape, y_train.shape)\n",
    "print(X_test_all.shape, y_test.shape)\n",
    "print(X_ext_test_all.shape, y_ext_test.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f528d2fd-3d56-4fe3-996c-872cba959650",
   "metadata": {},
   "source": [
    "# 4. Modeling\n",
    "time to train and evaluate!\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e40d529-602b-4bae-97e8-d527d48586f6",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 4.1. Single omics type: Methylation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2fb8f7f1-751e-40b0-9090-19c281392927",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "593\n",
      "(4863, 593) (4863,)\n",
      "(1239, 593) (1239,)\n",
      "(1842, 593) (1842,)\n"
     ]
    }
   ],
   "source": [
    "meth_columns = [col for col in df_multiomics.columns if 'cg' in col]  # all methylation columns (after pre-selection)\n",
    "# meth_columns = list(pd.read_csv('results/features/features_meth_top60.csv', usecols = ['Feature']).Feature) # selected features\n",
    "\n",
    "print(len(meth_columns))\n",
    "\n",
    "X_train = X_train_all[meth_columns]\n",
    "X_test = X_test_all[meth_columns]\n",
    "X_ext_test = X_ext_test_all[meth_columns]\n",
    "\n",
    "print(X_train.shape, y_train.shape)\n",
    "print(X_test.shape, y_test.shape)\n",
    "print(X_ext_test.shape, y_ext_test.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d266c429-d2f0-45af-a7bd-1c64dca911d7",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 3 folds for each of 4 candidates, totalling 12 fits\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:09] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:15] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:15] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:15] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:15] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:16] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:16] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:16] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:16] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:16] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:16] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:16] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:16] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:21] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:21] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:21] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:21] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:21] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:21] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:21] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:21] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:21] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:21] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:21] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:42:21] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best hyperparameters: {'colsample_bytree': 0.8, 'learning_rate': 0.02, 'max_depth': 5, 'n_estimators': 300, 'subsample': 0.8}\n",
      "========================================\n",
      "Test AUC: 0.9913\n",
      "========================================\n",
      "Time taken: 0.41 minutes\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "\n",
    "# Define the model\n",
    "model = XGBClassifier(objective='binary:logistic', tree_method='gpu_hist', gpu_id=0)\n",
    "\n",
    "# Define the hyperparameters for grid search\n",
    "param_grid = {\n",
    "    'n_estimators': [100, 300],\n",
    "    'max_depth': [5, ],\n",
    "    'learning_rate': [0.01, 0.02,],\n",
    "    'subsample': [0.8],\n",
    "    'colsample_bytree': [0.8],\n",
    "}\n",
    "\n",
    "# Set up GridSearchCV\n",
    "grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=3, scoring='roc_auc', verbose=1, n_jobs=-1)\n",
    "\n",
    "# Fit the model with cross-validation\n",
    "grid_search.fit(X_train, y_train)\n",
    "\n",
    "# Get the best model and hyperparameters\n",
    "best_model = grid_search.best_estimator_\n",
    "best_params = grid_search.best_params_\n",
    "\n",
    "print(f\"Best hyperparameters: {best_params}\")\n",
    "print(\"========================================\")\n",
    "\n",
    "# Train the best model on the full training set\n",
    "best_model.fit(X_train, y_train)\n",
    "\n",
    "# Make predictions on the test set\n",
    "y_pred_prob = best_model.predict_proba(X_test)[:, 1]  # Probabilities for the positive class\n",
    "\n",
    "# Calculate the AUC score\n",
    "auc = roc_auc_score(y_test, y_pred_prob)\n",
    "print(f\"Test AUC: {auc:.4f}\")\n",
    "print(\"========================================\")\n",
    "\n",
    "# Print execution time\n",
    "print(f\"Time taken: {(time.time() - start_time)/60:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "828df70a-1b76-4a12-85ca-d335ab93a2e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test AUC: 0.9813\n",
      "========================================\n",
      "Time taken: 0.41 minutes\n"
     ]
    }
   ],
   "source": [
    "y_ext_pred_prob = best_model.predict_proba(X_ext_test)[:, 1]  # Probabilities for the positive class\n",
    "\n",
    "# Calculate the AUC score\n",
    "auc = roc_auc_score(y_ext_test, y_ext_pred_prob)\n",
    "print(f\"Test AUC: {auc:.4f}\")\n",
    "print(\"========================================\")\n",
    "\n",
    "# Print execution time\n",
    "print(f\"Time taken: {(time.time() - start_time)/60:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "b778cef2-55bb-4391-96be-be945886dc16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved to results/features/features_meth_top60.csv\n"
     ]
    }
   ],
   "source": [
    "# feature selection\n",
    "num_top_features = 60\n",
    "importance = best_model.feature_importances_\n",
    "\n",
    "importance_df = pd.DataFrame({\n",
    "    'Feature': X_test.columns,\n",
    "    'Importance': importance\n",
    "})\n",
    "\n",
    "importance_df = importance_df.sort_values(by='Importance', ascending=False)\n",
    "\n",
    "top_meth_cols = importance_df.head(num_top_features)\n",
    "# top_meth_cols.to_csv(f'results/features/features_meth_top{num_top_features}.csv')\n",
    "print(f'Saved to results/features/features_meth_top{num_top_features}.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92416ded-fba6-4dcd-b0e8-7fcb48a5caaf",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## 4.2. Single omics type: gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "66ab968f-da79-478e-85ab-874f42785d01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "60660\n",
      "(4863, 60660) (4863,)\n",
      "(1239, 60660) (1239,)\n",
      "(1842, 60660) (1842,)\n"
     ]
    }
   ],
   "source": [
    "gene_exp_cols = [col for col in df_multiomics.columns if 'ENSG' in col] # all gene expression cols\n",
    "# gene_exp_cols = list(pd.read_csv('results/features/features_gene_exp_top60.csv', usecols = ['Feature']).Feature) # selected features\n",
    "\n",
    "print(len(gene_exp_cols))\n",
    "\n",
    "X_train = X_train_all[gene_exp_cols]\n",
    "X_test = X_test_all[gene_exp_cols]\n",
    "X_ext_test = X_ext_test_all[gene_exp_cols]\n",
    "\n",
    "print(X_train.shape, y_train.shape)\n",
    "print(X_test.shape, y_test.shape)\n",
    "print(X_ext_test.shape, y_ext_test.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f1dfa1da-7d42-418f-99e1-6ccc36c560c0",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 3 folds for each of 4 candidates, totalling 12 fits\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:46] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:46] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:47] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:48] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:48] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:48] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:48] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:48] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:43:48] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:44:12] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:44:13] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:44:47] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:44:47] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best hyperparameters: {'colsample_bytree': 0.8, 'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 100, 'subsample': 0.8}\n",
      "========================================\n",
      "Test AUC: 0.9917\n",
      "========================================\n",
      "Time taken: 3.80 minutes\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "\n",
    "# Define the model\n",
    "model = XGBClassifier(objective='binary:logistic', tree_method='gpu_hist', gpu_id=0)\n",
    "\n",
    "# Define the hyperparameters for grid search\n",
    "param_grid = {\n",
    "    'n_estimators': [100, 300],\n",
    "    'max_depth': [5, ],\n",
    "    'learning_rate': [0.01, 0.02,],\n",
    "    'subsample': [0.8],\n",
    "    'colsample_bytree': [0.8],\n",
    "}\n",
    "\n",
    "# Set up GridSearchCV\n",
    "grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=3, scoring='roc_auc', verbose=1, n_jobs=-1)\n",
    "\n",
    "# Fit the model with cross-validation\n",
    "grid_search.fit(X_train, y_train)\n",
    "\n",
    "# Get the best model and hyperparameters\n",
    "best_model = grid_search.best_estimator_\n",
    "best_params = grid_search.best_params_\n",
    "\n",
    "print(f\"Best hyperparameters: {best_params}\")\n",
    "print(\"========================================\")\n",
    "\n",
    "# Train the best model on the full training set\n",
    "best_model.fit(X_train, y_train)\n",
    "\n",
    "# Make predictions on the test set\n",
    "y_pred_prob = best_model.predict_proba(X_test)[:, 1]  # Probabilities for the positive class\n",
    "\n",
    "# Calculate the AUC score\n",
    "auc = roc_auc_score(y_test, y_pred_prob)\n",
    "print(f\"Test AUC: {auc:.4f}\")\n",
    "print(\"========================================\")\n",
    "\n",
    "# Print execution time\n",
    "print(f\"Time taken: {(time.time() - start_time)/60:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "8f3a61ef-b648-41ef-80f1-f1d13140c98f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test AUC: 0.9830\n",
      "========================================\n",
      "Time taken: 3.97 minutes\n"
     ]
    }
   ],
   "source": [
    "y_ext_pred_prob = best_model.predict_proba(X_ext_test)[:, 1]  # Probabilities for the positive class\n",
    "\n",
    "# Calculate the AUC score\n",
    "auc = roc_auc_score(y_ext_test, y_ext_pred_prob)\n",
    "print(f\"Test AUC: {auc:.4f}\")\n",
    "print(\"========================================\")\n",
    "\n",
    "# Print execution time\n",
    "print(f\"Time taken: {(time.time() - start_time)/60:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "d8dce4fd-02b7-49f7-be7c-162933f2808e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved to results/features/features_gene_exp_top60.csv\n"
     ]
    }
   ],
   "source": [
    "# feature selection\n",
    "num_top_features = 60\n",
    "importance = best_model.feature_importances_\n",
    "\n",
    "importance_df = pd.DataFrame({\n",
    "    'Feature': X_test.columns,\n",
    "    'Importance': importance\n",
    "})\n",
    "\n",
    "importance_df = importance_df.sort_values(by='Importance', ascending=False)\n",
    "\n",
    "top_gene_exp_cols = importance_df.head(num_top_features)\n",
    "# top_gene_exp_cols.to_csv(f'results/features/features_gene_exp_top{num_top_features}.csv')\n",
    "print(f'Saved to results/features/features_gene_exp_top{num_top_features}.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fe40db6-bed6-4927-9281-932c989f4dcd",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## 4.3. Single omics type: miRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "0936d1a3-1a7f-42af-b0db-6ad4e7e7e8f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1881\n",
      "(4863, 1881) (4863,)\n",
      "(1239, 1881) (1239,)\n",
      "(1842, 1881) (1842,)\n"
     ]
    }
   ],
   "source": [
    "mirna_cols = [col for col in df_multiomics.columns if 'hsa' in col]  # all miRNA columns\n",
    "# mirna_cols = list(pd.read_csv('results/features/features_mirna_top60.csv', usecols = ['Feature']).Feature) # selected features\n",
    "\n",
    "print(len(mirna_cols))\n",
    "\n",
    "X_train = X_train_all[mirna_cols]\n",
    "X_test = X_test_all[mirna_cols]\n",
    "X_ext_test = X_ext_test_all[mirna_cols]\n",
    "\n",
    "print(X_train.shape, y_train.shape)\n",
    "print(X_test.shape, y_test.shape)\n",
    "print(X_ext_test.shape, y_ext_test.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "efff1107-3179-47a2-b1b4-c64c2f8336cc",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 3 folds for each of 4 candidates, totalling 12 fits\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:26] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:27] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:27] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:27] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:27] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:28] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:28] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:34] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:34] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:34] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:34] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:34] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:34] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:34] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:34] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:34] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:34] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:40] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:40] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:41] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:41] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:41] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:41] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:41] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:41] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:41] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:41] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:41] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:46:41] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best hyperparameters: {'colsample_bytree': 0.8, 'learning_rate': 0.02, 'max_depth': 5, 'n_estimators': 300, 'subsample': 0.8}\n",
      "========================================\n",
      "Test AUC: 0.9907\n",
      "========================================\n",
      "Time taken: 0.39 minutes\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "\n",
    "# Define the model\n",
    "model = XGBClassifier(objective='binary:logistic', tree_method='gpu_hist', gpu_id=0)\n",
    "\n",
    "# Define the hyperparameters for grid search\n",
    "param_grid = {\n",
    "    'n_estimators': [100, 300],\n",
    "    'max_depth': [5, ],\n",
    "    'learning_rate': [0.01, 0.02,],\n",
    "    'subsample': [0.8],\n",
    "    'colsample_bytree': [0.8],\n",
    "}\n",
    "\n",
    "# Set up GridSearchCV\n",
    "grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=3, scoring='roc_auc', verbose=1, n_jobs=-1)\n",
    "\n",
    "# Fit the model with cross-validation\n",
    "grid_search.fit(X_train, y_train)\n",
    "\n",
    "# Get the best model and hyperparameters\n",
    "best_model = grid_search.best_estimator_\n",
    "best_params = grid_search.best_params_\n",
    "\n",
    "print(f\"Best hyperparameters: {best_params}\")\n",
    "print(\"========================================\")\n",
    "\n",
    "# Train the best model on the full training set\n",
    "best_model.fit(X_train, y_train)\n",
    "\n",
    "# Make predictions on the test set\n",
    "y_pred_prob = best_model.predict_proba(X_test)[:, 1]  # Probabilities for the positive class\n",
    "\n",
    "# Calculate the AUC score\n",
    "auc = roc_auc_score(y_test, y_pred_prob)\n",
    "print(f\"Test AUC: {auc:.4f}\")\n",
    "print(\"========================================\")\n",
    "\n",
    "# Print execution time\n",
    "print(f\"Time taken: {(time.time() - start_time)/60:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "8091278c-7955-4fdd-9f41-15240465be87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test AUC: 0.9420\n",
      "========================================\n",
      "Time taken: 0.39 minutes\n"
     ]
    }
   ],
   "source": [
    "y_ext_pred_prob = best_model.predict_proba(X_ext_test)[:, 1]  # Probabilities for the positive class\n",
    "\n",
    "# Calculate the AUC score\n",
    "auc = roc_auc_score(y_ext_test, y_ext_pred_prob)\n",
    "print(f\"Test AUC: {auc:.4f}\")\n",
    "print(\"========================================\")\n",
    "\n",
    "# Print execution time\n",
    "print(f\"Time taken: {(time.time() - start_time)/60:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "cd343ea3-70b9-410b-9815-d251b632a2bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "results/features/features_mirna_top60.csv\n"
     ]
    }
   ],
   "source": [
    "# feature selection\n",
    "num_top_features = 60\n",
    "importance = best_model.feature_importances_\n",
    "\n",
    "importance_df = pd.DataFrame({\n",
    "    'Feature': X_test.columns,\n",
    "    'Importance': importance\n",
    "})\n",
    "\n",
    "importance_df = importance_df.sort_values(by='Importance', ascending=False)\n",
    "\n",
    "top_mirna_cols = importance_df.head(num_top_features)\n",
    "# top_mirna_cols.to_csv(f'results/features/features_mirna_top{num_top_features}.csv')\n",
    "print(f'results/features/features_mirna_top{num_top_features}.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "080dd50f-1c60-406c-acf3-0b7a50d7b00c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## 4.4. Multiomics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "7157a8c3-0bd9-4310-8890-f78c88b9ad42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to count number features per each omics included \n",
    "def count_features(df_column):\n",
    "    methylation_count = 0\n",
    "    gene_exp_count = 0\n",
    "    miRNA_count = 0\n",
    "\n",
    "    sample_count = 0\n",
    "    site_count = 0\n",
    "    stage_count = 0\n",
    "    type_count = 0\n",
    "\n",
    "    overall_count = 0\n",
    "\n",
    "    for col in df_column:\n",
    "        overall_count += 1\n",
    "        if 'ENSG' in col:\n",
    "            gene_exp_count += 1\n",
    "\n",
    "        if 'hsa' in col:\n",
    "            miRNA_count += 1\n",
    "\n",
    "        if 'cg' in col:\n",
    "            methylation_count += 1\n",
    "    col_counts = pd.DataFrame()\n",
    "    col_counts['col_type'] = ['Methylation', 'Gene_Expression', 'miRNA']\n",
    "    col_counts['col_type_count'] = [methylation_count, gene_exp_count, miRNA_count]\n",
    "    # print(overall_count)\n",
    "    # print(sum(col_counts.col_type_count) == overall_count)\n",
    "    return col_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "df954bba-b801-4328-85c1-27007580dbf6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "91"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# prevent overload\n",
    "import gc\n",
    "del model\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "81f26a08-94b5-468d-b2b0-929b2b7c03bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "63134\n",
      "(4863, 63134) (4863,)\n",
      "(1239, 63134) (1239,)\n",
      "(1842, 63134) (1842,)\n"
     ]
    }
   ],
   "source": [
    "## 3 omics types (gene expression + miRNA + methylation)\n",
    "selected_cols = [x for x in df_multiomics.columns if ('ENSG' in x) or ('hsa' in x) or ('cg' in x)] # include all features in all omics types \n",
    "# selected_cols = list(pd.read_csv('results/features/features_tri_omics_top1500_overall.csv', usecols = ['Feature']).Feature) # selected features \n",
    "# selected_cols = list(pd.read_csv('results/features/features_tri_omics_top500_per_prefix.csv', usecols = ['Feature']).Feature) # selected features \n",
    "\n",
    "\n",
    "## 2 omics types (gene expression + methylation)\n",
    "# selected_cols = [x for x in df_multiomics.columns if ('ENSG' in x) or ('cg' in x)] # all features in the 2 selected omics types (gene expression + methylation)\n",
    "# selected_cols = list(pd.read_csv('results/features/features_bi_omics_gene_and_meth_top1000_overall.csv', usecols = ['Feature']).Feature) # selected features \n",
    "\n",
    "print(len(selected_cols))\n",
    "\n",
    "X_train = X_train_all[selected_cols]\n",
    "X_test = X_test_all[selected_cols]\n",
    "X_ext_test = X_ext_test_all[selected_cols]\n",
    "\n",
    "print(X_train.shape, y_train.shape)\n",
    "print(X_test.shape, y_test.shape)\n",
    "print(X_ext_test.shape, y_ext_test.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "d50e871a-6f4f-46ec-9b99-106a8d0df4cb",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 3 folds for each of 6 candidates, totalling 18 fits\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:27] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:27] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:27] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:27] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:28] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:28] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:28] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:28] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:28] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:28] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:28] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:28] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:29] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:30] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:30] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:30] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:30] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:30] WARNING: /workspace/src/common/error_msg.cc:45: `gpu_id` is deprecated since2.0.0, use `device` instead. E.g. device=cpu/cuda/cuda:0\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:48:30] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:50:15] WARNING: /workspace/src/common/error_msg.cc:27: The tree method `gpu_hist` is deprecated since 2.0.0. To use GPU training, set the `device` parameter to CUDA instead.\n",
      "\n",
      "    E.g. tree_method = \"hist\", device = \"cuda\"\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n",
      "/home/gridsan/cwang1/.local/lib/python3.9/site-packages/xgboost/core.py:158: UserWarning: [13:50:15] WARNING: /workspace/src/common/error_msg.cc:58: Falling back to prediction using DMatrix due to mismatched devices. This might lead to higher memory usage and slower performance. XGBoost is running on: cuda:0, while the input data is on: cpu.\n",
      "Potential solutions:\n",
      "- Use a data structure that matches the device ordinal in the booster.\n",
      "- Set the device for booster before call to inplace_predict.\n",
      "\n",
      "This warning will only be shown once.\n",
      "\n",
      "  warnings.warn(smsg, UserWarning)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: black;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-1 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-1 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: block;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"▸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content {\n",
       "  max-height: 0;\n",
       "  max-width: 0;\n",
       "  overflow: hidden;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  max-height: 200px;\n",
       "  max-width: 100%;\n",
       "  overflow: auto;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"▾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-1 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-1 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-1 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 1ex;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-1 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>GridSearchCV(cv=3,\n",
       "             estimator=XGBClassifier(base_score=None, booster=None,\n",
       "                                     callbacks=None, colsample_bylevel=None,\n",
       "                                     colsample_bynode=None,\n",
       "                                     colsample_bytree=None, device=None,\n",
       "                                     early_stopping_rounds=None,\n",
       "                                     enable_categorical=False, eval_metric=None,\n",
       "                                     feature_types=None, gamma=None, gpu_id=0,\n",
       "                                     grow_policy=None, importance_type=None,\n",
       "                                     interaction_constraints=None,\n",
       "                                     learning_r...\n",
       "                                     max_delta_step=None, max_depth=None,\n",
       "                                     max_leaves=None, min_child_weight=None,\n",
       "                                     missing=nan, monotone_constraints=None,\n",
       "                                     multi_strategy=None, n_estimators=None,\n",
       "                                     n_jobs=None, num_parallel_tree=None, ...),\n",
       "             n_jobs=-1,\n",
       "             param_grid={&#x27;colsample_bytree&#x27;: [0.8],\n",
       "                         &#x27;learning_rate&#x27;: [0.01, 0.02, 0.05], &#x27;max_depth&#x27;: [5],\n",
       "                         &#x27;n_estimators&#x27;: [100, 300], &#x27;subsample&#x27;: [0.8]},\n",
       "             scoring=&#x27;roc_auc&#x27;, verbose=1)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item sk-dashed-wrapped\"><div class=\"sk-label-container\"><div class=\"sk-label fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" ><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">&nbsp;&nbsp;GridSearchCV<a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.4/modules/generated/sklearn.model_selection.GridSearchCV.html\">?<span>Documentation for GridSearchCV</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></label><div class=\"sk-toggleable__content fitted\"><pre>GridSearchCV(cv=3,\n",
       "             estimator=XGBClassifier(base_score=None, booster=None,\n",
       "                                     callbacks=None, colsample_bylevel=None,\n",
       "                                     colsample_bynode=None,\n",
       "                                     colsample_bytree=None, device=None,\n",
       "                                     early_stopping_rounds=None,\n",
       "                                     enable_categorical=False, eval_metric=None,\n",
       "                                     feature_types=None, gamma=None, gpu_id=0,\n",
       "                                     grow_policy=None, importance_type=None,\n",
       "                                     interaction_constraints=None,\n",
       "                                     learning_r...\n",
       "                                     max_delta_step=None, max_depth=None,\n",
       "                                     max_leaves=None, min_child_weight=None,\n",
       "                                     missing=nan, monotone_constraints=None,\n",
       "                                     multi_strategy=None, n_estimators=None,\n",
       "                                     n_jobs=None, num_parallel_tree=None, ...),\n",
       "             n_jobs=-1,\n",
       "             param_grid={&#x27;colsample_bytree&#x27;: [0.8],\n",
       "                         &#x27;learning_rate&#x27;: [0.01, 0.02, 0.05], &#x27;max_depth&#x27;: [5],\n",
       "                         &#x27;n_estimators&#x27;: [100, 300], &#x27;subsample&#x27;: [0.8]},\n",
       "             scoring=&#x27;roc_auc&#x27;, verbose=1)</pre></div> </div></div><div class=\"sk-parallel\"><div class=\"sk-parallel-item\"><div class=\"sk-item\"><div class=\"sk-label-container\"><div class=\"sk-label fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-2\" type=\"checkbox\" ><label for=\"sk-estimator-id-2\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">estimator: XGBClassifier</label><div class=\"sk-toggleable__content fitted\"><pre>XGBClassifier(base_score=None, booster=None, callbacks=None,\n",
       "              colsample_bylevel=None, colsample_bynode=None,\n",
       "              colsample_bytree=None, device=None, early_stopping_rounds=None,\n",
       "              enable_categorical=False, eval_metric=None, feature_types=None,\n",
       "              gamma=None, gpu_id=0, grow_policy=None, importance_type=None,\n",
       "              interaction_constraints=None, learning_rate=None, max_bin=None,\n",
       "              max_cat_threshold=None, max_cat_to_onehot=None,\n",
       "              max_delta_step=None, max_depth=None, max_leaves=None,\n",
       "              min_child_weight=None, missing=nan, monotone_constraints=None,\n",
       "              multi_strategy=None, n_estimators=None, n_jobs=None,\n",
       "              num_parallel_tree=None, ...)</pre></div> </div></div><div class=\"sk-serial\"><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-3\" type=\"checkbox\" ><label for=\"sk-estimator-id-3\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">XGBClassifier</label><div class=\"sk-toggleable__content fitted\"><pre>XGBClassifier(base_score=None, booster=None, callbacks=None,\n",
       "              colsample_bylevel=None, colsample_bynode=None,\n",
       "              colsample_bytree=None, device=None, early_stopping_rounds=None,\n",
       "              enable_categorical=False, eval_metric=None, feature_types=None,\n",
       "              gamma=None, gpu_id=0, grow_policy=None, importance_type=None,\n",
       "              interaction_constraints=None, learning_rate=None, max_bin=None,\n",
       "              max_cat_threshold=None, max_cat_to_onehot=None,\n",
       "              max_delta_step=None, max_depth=None, max_leaves=None,\n",
       "              min_child_weight=None, missing=nan, monotone_constraints=None,\n",
       "              multi_strategy=None, n_estimators=None, n_jobs=None,\n",
       "              num_parallel_tree=None, ...)</pre></div> </div></div></div></div></div></div></div></div></div>"
      ],
      "text/plain": [
       "GridSearchCV(cv=3,\n",
       "             estimator=XGBClassifier(base_score=None, booster=None,\n",
       "                                     callbacks=None, colsample_bylevel=None,\n",
       "                                     colsample_bynode=None,\n",
       "                                     colsample_bytree=None, device=None,\n",
       "                                     early_stopping_rounds=None,\n",
       "                                     enable_categorical=False, eval_metric=None,\n",
       "                                     feature_types=None, gamma=None, gpu_id=0,\n",
       "                                     grow_policy=None, importance_type=None,\n",
       "                                     interaction_constraints=None,\n",
       "                                     learning_r...\n",
       "                                     max_delta_step=None, max_depth=None,\n",
       "                                     max_leaves=None, min_child_weight=None,\n",
       "                                     missing=nan, monotone_constraints=None,\n",
       "                                     multi_strategy=None, n_estimators=None,\n",
       "                                     n_jobs=None, num_parallel_tree=None, ...),\n",
       "             n_jobs=-1,\n",
       "             param_grid={'colsample_bytree': [0.8],\n",
       "                         'learning_rate': [0.01, 0.02, 0.05], 'max_depth': [5],\n",
       "                         'n_estimators': [100, 300], 'subsample': [0.8]},\n",
       "             scoring='roc_auc', verbose=1)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "\n",
    "# Define the model\n",
    "model = XGBClassifier(objective='binary:logistic', tree_method='gpu_hist', gpu_id=0)\n",
    "\n",
    "# Define the hyperparameters for grid search\n",
    "param_grid = {\n",
    "    'n_estimators': [100, 300],\n",
    "    'max_depth': [5, ],\n",
    "    'learning_rate': [0.01, 0.02, 0.05],\n",
    "    'subsample': [0.8],\n",
    "    'colsample_bytree': [0.8],\n",
    "}\n",
    "\n",
    "# Set up GridSearchCV\n",
    "grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=3, scoring='roc_auc', verbose=1, n_jobs=-1)\n",
    "\n",
    "# Fit the model with cross-validation\n",
    "grid_search.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "68255cca-9966-48e0-a888-cb689856f233",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best hyperparameters: {'colsample_bytree': 0.8, 'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 100, 'subsample': 0.8}\n",
      "========================================\n",
      "Test AUC: 0.9909\n",
      "========================================\n",
      "Time taken: 6.04 minutes\n"
     ]
    }
   ],
   "source": [
    "# Get the best model and hyperparameters\n",
    "best_model = grid_search.best_estimator_\n",
    "best_params = grid_search.best_params_\n",
    "\n",
    "print(f\"Best hyperparameters: {best_params}\")\n",
    "print(\"========================================\")\n",
    "\n",
    "# Train the best model on the full training set\n",
    "best_model.fit(X_train, y_train)\n",
    "\n",
    "# Make predictions on the test set\n",
    "y_pred_prob = best_model.predict_proba(X_test)[:, 1]  # Probabilities for the positive class\n",
    "\n",
    "# Calculate the AUC score\n",
    "auc = roc_auc_score(y_test, y_pred_prob)\n",
    "print(f\"Test AUC: {auc:.4f}\")\n",
    "print(\"========================================\")\n",
    "\n",
    "# Print execution time\n",
    "print(f\"Time taken: {(time.time() - start_time)/60:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "2392994c-2652-4f62-8432-fd2714dab2bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Test AUC: 0.9805\n",
      "========================================\n",
      "Time taken: 6.21 minutes\n"
     ]
    }
   ],
   "source": [
    "y_ext_pred_prob = best_model.predict_proba(X_ext_test)[:, 1]  # Probabilities for the positive class\n",
    "\n",
    "# Calculate the AUC score\n",
    "auc = roc_auc_score(y_ext_test, y_ext_pred_prob)\n",
    "print(f\"Test AUC: {auc:.4f}\")\n",
    "print(\"========================================\")\n",
    "\n",
    "print(f\"Time taken: {(time.time() - start_time)/60:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "bc0a0822-b435-4065-aa42-7c9ea4268c12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>col_type</th>\n",
       "      <th>col_type_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Methylation</td>\n",
       "      <td>593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gene_Expression</td>\n",
       "      <td>60660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>miRNA</td>\n",
       "      <td>1881</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          col_type  col_type_count\n",
       "0      Methylation             593\n",
       "1  Gene_Expression           60660\n",
       "2            miRNA            1881"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_features(X_train.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ec82293-b80f-41f7-84ac-dcad77112e62",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 4.4.1. Feature Selection Approach 1 (Select top n overall)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "5714ceb1-94ab-4296-822e-fb4471e67483",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved to results/features/features_tri_omics_gene_and_meth_top1500_overall.csv\n"
     ]
    }
   ],
   "source": [
    "importance = best_model.feature_importances_\n",
    "num_top_features = 1500 \n",
    "\n",
    "importance_df = pd.DataFrame({\n",
    "    'Feature': X_test.columns,\n",
    "    'Importance': importance\n",
    "})\n",
    "\n",
    "importance_df = importance_df.sort_values(by='Importance', ascending=False)\n",
    "\n",
    "top_multiomics_cols = importance_df.head(num_top_features)\n",
    "\n",
    "# top_multiomics_cols.to_csv(f'results/features/features_tri_omics_gene_and_meth_top{num_top_features}_overall.csv')\n",
    "print(f\"Saved to results/features/features_tri_omics_gene_and_meth_top{num_top_features}_overall.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "a36dabb6-c4ac-48db-98b3-0c913d990b4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>col_type</th>\n",
       "      <th>col_type_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Methylation</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gene_Expression</td>\n",
       "      <td>1372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>miRNA</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          col_type  col_type_count\n",
       "0      Methylation              86\n",
       "1  Gene_Expression            1372\n",
       "2            miRNA              42"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_features(list(top_multiomics_cols.Feature.values))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b8b4d00-5b22-4858-96de-548fd9a8215d",
   "metadata": {},
   "source": [
    "### 4.4.2. Feature Selection Approach 2 (Balanced: Select top k per omics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "7eec8b23-306f-4420-95ea-bdd7c8ce28a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(63134, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>Importance</th>\n",
       "      <th>prefix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10641</th>\n",
       "      <td>ENSG00000161888.11</td>\n",
       "      <td>0.033914</td>\n",
       "      <td>ENSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60942</th>\n",
       "      <td>hsa-mir-21</td>\n",
       "      <td>0.021350</td>\n",
       "      <td>hsa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3427</th>\n",
       "      <td>ENSG00000107562.16</td>\n",
       "      <td>0.019378</td>\n",
       "      <td>ENSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6793</th>\n",
       "      <td>ENSG00000133800.9</td>\n",
       "      <td>0.017456</td>\n",
       "      <td>ENSG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62735</th>\n",
       "      <td>cg00808170</td>\n",
       "      <td>0.016653</td>\n",
       "      <td>cg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62872</th>\n",
       "      <td>cg00379708</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62873</th>\n",
       "      <td>cg02910987</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62875</th>\n",
       "      <td>cg00006081</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62877</th>\n",
       "      <td>cg00932063</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62878</th>\n",
       "      <td>cg17766828</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cg</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1500 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Feature  Importance prefix\n",
       "10641  ENSG00000161888.11    0.033914   ENSG\n",
       "60942          hsa-mir-21    0.021350    hsa\n",
       "3427   ENSG00000107562.16    0.019378   ENSG\n",
       "6793    ENSG00000133800.9    0.017456   ENSG\n",
       "62735          cg00808170    0.016653     cg\n",
       "...                   ...         ...    ...\n",
       "62872          cg00379708    0.000000     cg\n",
       "62873          cg02910987    0.000000     cg\n",
       "62875          cg00006081    0.000000     cg\n",
       "62877          cg00932063    0.000000     cg\n",
       "62878          cg17766828    0.000000     cg\n",
       "\n",
       "[1500 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>col_type</th>\n",
       "      <th>col_type_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Methylation</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gene_Expression</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>miRNA</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          col_type  col_type_count\n",
       "0      Methylation             500\n",
       "1  Gene_Expression             500\n",
       "2            miRNA             500"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved to results/features/features_tri_omics_top500_per_prefix.csv\n"
     ]
    }
   ],
   "source": [
    "importance = best_model.feature_importances_\n",
    "importance_df = pd.DataFrame({\n",
    "    'Feature': X_test.columns,\n",
    "    'Importance': importance\n",
    "})\n",
    "importance_df = importance_df.sort_values(by='Importance', ascending=False)\n",
    "importance_df['prefix'] = importance_df['Feature'].str.extract(r'^(ENSG|cg|hsa)')\n",
    "print(importance_df.shape)\n",
    "\n",
    "n_features_per_omics = 500\n",
    "\n",
    "top_features_per_prefix = importance_df.groupby('prefix').head(n_features_per_omics)\n",
    "\n",
    "display(top_features_per_prefix)\n",
    "display(count_features(list(top_features_per_prefix.Feature.values)))\n",
    "\n",
    "# top_features_per_prefix.to_csv(f'results/features/features_tri_omics_top{n_features_per_omics}_per_prefix.csv')\n",
    "print(f\"Saved to results/features/features_tri_omics_top{n_features_per_omics}_per_prefix.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fc374be-0d63-4602-a9ad-929c502f4f5d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
